KEP-5007: DRA Device Binding Conditions (#5012)
kubernetes/enhancements: keps/sig-scheduling/5007-device-attach-before-pod-scheduled/README.md

    //
    // +optional
    // +oneOf=deviceType
    Basic *BasicDevice
}

// BasicDevice represents a basic device instance.
type BasicDevice struct {
    // Attributes contains additional attributes of the device.
    //
    // +optional
    Attributes map[string]string
}
```

To indicate a fabric device, the following attribute will be added:

```yaml
attributes:
  kubernetes.io/needs-attaching:
    bool: "true"
```

#### `AllocatedDeviceStatus` Additions

The `Conditions` field within `AllocatedDeviceStatus` is used to indicate the status of the device attachment.
This field will contain a list of conditions, each representing a specific state or event related to the device.

For this feature, the NodeName and following `ConditionType` constants are added:

```go
// AllocatedDeviceStatus contains the status of an allocated device, if the
// driver chooses to report it. This may include driver-specific information.
type AllocatedDeviceStatus struct {
...
    // NodeName contains the name of the node where the device needs to be attached.
    //
    // +optional
    NodeName string 
}
const(
  DRADeviceNeedAttachType = "kubernetes.io/needs-attaching"

>   DRADeviceIsAttachType = "kubernetes.io/is-attached"

Author: If nothing is done, I believe the claim would remain in the is-attached state indefinitely. However, I expected that in DRA, the `ResourceClaim` would be deleted when the pod is removed. This means that the lifespan of the `ResourceClaim` would be almost the same as that of the pod.

In cases where binding fails for reasons other than DRA, I think the unbinding would be handled by the UnReserve operation.

@pohly, could you please share your thoughts on the necessity of a timeout in this context?

>   DRADeviceIsAttachType = "kubernetes.io/is-attached"

Reviewer: A device might get attached for a claim which is currently marked as allocated. This gets recorded in the condition of that device in the claim status.

Then if anything fails, like not being able to actually schedule the pod to the node, the same recovery mechanism as in other failure scenarios kicks in: the allocation is removed and the scheduler tries to allocate devices for some other node. This happens in PostFilter if it was determined that an allocated claim forced the pod onto a node and that selection didn't work out.

It doesn't matter whether the device is attached or not. The DRA driver will at some point see a request to attach it elsewhere and then should do that.

The DRA driver may also do some "garbage collection": if a device is attached without being allocated, then it's certain that the device is not needed anymore on the node. This can be useful because detaching a device is never explicitly requested.

  DRADeviceAttachFailType = "kubernetes.io/attach-failed"
)
```

When `kubernetes.io/needs-attaching: true` is set, the scheduler is expected to do the following:

1. Set `AllocatedDeviceStatus.NodeName`.
2. Add an `AllocatedDeviceStatus` with a condition of `Type: kubernetes.io/needs-attaching` and `Status: True`.

> 3. Wait for a condition with `Type: kubernetes.io/is-attached` and `Status: True` in `PreBind` before proceeding.

Author: We intend to implement a timeout for waiting in the PreBind phase. I will make sure to specify this clearly in the documentation.

4. Give up when seeing a condition with `Type: kubernetes.io/attach-failed` and `Status: True`.

#### Handling ResourceSlices Upon Failure of Attachment

During the scheduling cycle, the DRA plugin reserves a `ResourceSlice` for the `ResourceClaim`.
In the binding cycle, the reserved `ResourceSlice` is assigned during `PreBind`.

If a fabric device is selected, the scheduler waits for the device attachment during `PreBind`.
The composable controller performs the attachment operation by checking the flag of the `ResourceClaim`.
If the attachment fails, the following steps are taken:

> 1. **Update ResourceClaim**: The composable controller updates the `ResourceClaim` to indicate the failure of the attachment by setting a condition with `Type: kubernetes.io/is-attached` and `Status: False`.

Reviewer 2: ```suggestion
1. **Update ResourceClaim**: The composable controller updates the `AllocatedDeviceStatus` to indicate the failure of the attachment by setting a condition with `Type: kubernetes.io/is-attached` and `Status: False`.
```

> 1. **Update ResourceClaim**: The composable controller updates the `ResourceClaim` to indicate the failure of the attachment by setting a condition with `Type: kubernetes.io/is-attached` and `Status: False`.

Reviewer 2: Isn't the `kubernetes.io/is-attached` False by default? Did you meant it sets `kubernetes.io/attach-failed` to True instead?

Author: You are correct. In this case, compostable controller set `Status: True` for `Type: kubernetes.io/attach-failed`. I will correct this mistake. Thank you for pointing it out.

2. **Fail the Binding Cycle**: The scheduler detects the failed attachment condition and fails the binding cycle. This prevents the pod from proceeding with an unattached device.

> 3. **Unbind ResourceClaim and ResourceSlice**: The scheduler unbinds the `ResourceClaim` and `ResourceSlice`, clearing the allocation to prevent the fabric device from being used in the pool.

Reviewer 2: Maybe it's obvious, but is it a requirement that the resource claim is ephemeral (dedicated to a given pod)? Otherwise unbinding the whole resource claim would affect other pods.

Author: Thank you for your insightful question. I expected that the DRA `ResourceClaim` is ephemeral and dedicated to a given pod to ensure that unbinding the ResourceClaim and ResourceSlice will not affect other pods. However, in cases where multiple pods share a ResourceClaim, we need to handle this carefully. For instance, we should ensure that unbinding only affects the device requested by the currently scheduled pod or the device that has failed.

To illustrate, if Pod A and Pod B share a `ResourceClaim` and Pod A's device attachment fails, we should only unbind the device associated with Pod A without affecting Pod B's allocation.

@pohly, could you please clarify which cases we need to consider here to ensure we handle shared `ResourceClaims` appropriately?

Reviewer: When a claim is shared between two pods, there is no "pod A's device" - it is the same device for both pods.

That means that device needs to be attached to a node and must remain attached until all pods have terminated. At that point, the claim will get deallocated and the driver controller can detach it or wait until it needs to be attached elsewhere.

Attaching has to complete before binding the first pod (the point of this KEP). By keeping the claim "in flight" (there's a list of those in the scheduler plugin), no other pod sharing that claim gets scheduled, so they also wait.

Author: Thank you for the explanation. It seems I had a slight misunderstanding about shared `ResourceClaims`. Based on the explanation, I believe that unbinding in the event of an attachment failure should not cause any issues.

Reviewer: Agreed, that is also my understanding.

> 3. **Unbind ResourceClaim and ResourceSlice**: The scheduler unbinds the `ResourceClaim` and `ResourceSlice`, clearing the allocation to prevent the fabric device from being used in the pool.

Reviewer 2: ```suggestion
3. **Unbind ResourceClaim and ResourceSlice**: The scheduler DRA plugin unbinds the `ResourceClaim` and `ResourceSlice`, clearing the allocation to prevent the fabric device from being used in the pool.
```

- [ ] (R) Production readiness review completed
- [ ] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

To achieve efficient management of fabric devices, we propose adding the following features to the Kubernetes scheduler's DRA plugin.
Fabric devices are those that are not directly connected to the server and require attachment to the server for use.

> In the DRA current implementation, fabric devices are attached after the scheduling decision, which leads to the following issue:

Reviewer 3: ```suggestion
In the current DRA implementation, fabric devices are attached after the scheduling decision, which leads to the following issue:
```

Fabric devices may be contested by other clusters.
In scenarios where attachment occurs after scheduling, there is a risk that the resource cannot be attached at the time of attachment, causing the container to remain in the "Container Creating" state.

To address this issue, we propose a feature that allows the DRA scheduler plugin to wait for the device to be attached.
If attaching fails, the scheduler will retry and attempt to schedule the pod elsewhere.

## Motivation

As AI and ML become popular in container (K8s) environment, enormous computational resources are required more and more.
On the other hand, efforts toward energy efficiency are also required for the realization of a sustainable society.
It is expected to achieve the conflicting requirements that providing higher performance and reducing power consumption simultaneously.
Recently, a new server architecture called Composable Disaggregated Infrastructure is emerged.

In a traditional server, hardware resources such as CPUs, memory, and GPUs reside within the server.
Composable Disaggregated Infrastructure decomposes these hardware resources and makes them available as resource pools.

> We can combine these resource by software definition so that we can create custom-made servers.

Reviewer 3: ```suggestion
We can combine these resources by software definition so that we can create custom-made servers.
```

This approach is superior because it avoids unnecessary waiting and allows for immediate rescheduling.

### Goals

1. **Enhance the DRA Scheduling Process**: 
Implement a feature that allows the scheduling process to wait for the completion of fabric device attachment.
This ensures that pods are only scheduled once the necessary fabric devices are successfully attached, improving reliability and efficiency.

2. **Attribute Information for Fabric Devices**: 
Add attribute information that clearly distinguishes fabric devices requiring attachment.
This will help in accurately identifying and managing these devices within the Kubernetes environment.

3. **Prioritize Device Allocation**:
Implement a prioritization mechanism for device allocation, favoring devices directly connected to the node over attached fabric devices.
This hierarchy ensures optimal performance and resource utilization.
For example, the order of preference would be: Node-local devices > Attached fabric devices > Pre-attached fabric devices.

### Non-Goals

<!--
What is out of scope for this KEP? Listing non-goals helps to focus discussion
and make progress.
-->

## Proposal

The basic idea is the following:

> 1. **Add a Ready Flag to ResourceClaim**:

Reviewer 3: Do we need a diagram in this doc to show how "Ready" state transition with other possible states.

   - Add a flag to `ResourceClaim` that indicates the readiness state of the device.

>    The `PreBind` phase will be held until this flag is set to "Ready".

Reviewer 4: Who sets this state in cases when it is a non-fabric attached device?

Reviewer: I think we moved away from this API already - see https://github.com/kubernetes/enhancements/pull/5012#discussion_r1937414834

Reviewer 4: Ah -- I should have read read the comments before jumping into the text.

bogged down.
-->

#### Story 1

#### Story 2

### Notes/Constraints/Caveats (Optional)

<!--
What are the caveats to the proposal?
What are some important details that didn't come across above?
Go in to as much detail as necessary here.
This might be a good place to talk about core concepts and how they relate.
-->

### Risks and Mitigations

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

## Design Details

### DRA Scheduler Plugin Design Overview

This document outlines the design of the DRA Scheduler Plugin, focusing on the handling of fabric devices.
Key additions include new attributes for device identification, enhancements to `AllocatedDeviceStatus`, and the process for handling `ResourceSlices` upon attachment failure.
The composable controller design is also discussed, emphasizing efficient utilization of fabric devices.

//TODO update figure

> ![proposal](proposal2.JPG)

Reviewer 3: This figure didn't show the timeout case (10 mins you mentioned in the following section).

#### Device Attribute Additions

To indicate whether a device is a fabric device, an attribute is added to the `Basic` within `Device`.
This attribute will be used by the controller that exposes the `ResourceSlice` to notify whether the device is a fabric device.
To avoid impacting existing DRA functionality, the default value of this attribute is set to `false`.

```go
// Device represents one individual hardware instance that can be selected based
// on its attributes. Besides the name, exactly one field must be set.
type Device struct {
    // Name is unique identifier among all devices managed by
    // the driver in the pool. It must be a DNS label.
    //
    // +required
    Name string

    // Basic defines one device instance.
    //
    // +optional
    // +oneOf=deviceType
    Basic *BasicDevice
}

// BasicDevice represents a basic device instance.
type BasicDevice struct {
    // Attributes contains additional attributes of the device.
    //
    // +optional
    Attributes map[string]string
}
```

To indicate a fabric device, the following attribute will be added:

```yaml
attributes:
  kubernetes.io/needs-attaching:

>     bool: "true"

Reviewer 4: I'm not sure about the naming of this attribute. I get its purpose, but the current name seems awkward.

```

#### `AllocatedDeviceStatus` Additions

The `Conditions` field within `AllocatedDeviceStatus` is used to indicate the status of the device attachment.
This field will contain a list of conditions, each representing a specific state or event related to the device.

For this feature, the NodeName and following `ConditionType` constants are added:

```go
// AllocatedDeviceStatus contains the status of an allocated device, if the
// driver chooses to report it. This may include driver-specific information.
type AllocatedDeviceStatus struct {
...
    // NodeName contains the name of the node where the device needs to be attached.
    //
    // +optional
    NodeName string 
}

> const(

Reviewer 5: we don't need any of these pre-defined in k8s, if we go with the binding gates approach. These all become vendor specific content published by the driver.

The DRA machinery just needs to:
- look for the gates, if they are there then look at the conditions and wait until they are met
- look for the failure conditions and check for failure while waiting
- have some timeout value (I think it should also be in the ResourceSlice)

Author: As you point out, a conditional approach is preferable for DRA, so I will update the KEP to reflect that we are using a binding gate approach.

Author: It seems like it needs some more discussion, so I'll update the KEP later.

  DRADeviceNeedAttachType = "kubernetes.io/needs-attaching"
  DRADeviceIsAttachType = "kubernetes.io/is-attached"
  DRADeviceAttachFailType = "kubernetes.io/attach-failed"
)
```

When `kubernetes.io/needs-attaching: true` is set, the scheduler DRA plugin is expected to do the following:

1. Set `AllocatedDeviceStatus.NodeName`.
2. Add an `AllocatedDeviceStatus` with a condition of `Type: kubernetes.io/needs-attaching` and `Status: True`.
3. Wait for a condition with `Type: kubernetes.io/is-attached` and `Status: True` in `PreBind` before proceeding.
4. Give up when seeing a condition with `Type: kubernetes.io/attach-failed` and `Status: True`.

Note: There is a concern that the in-flight events cache may grow too large when waiting in PreBind.
This issue will be addressed separately as outlined in kubernetes/kubernetes#129967.

#### PreBind Phase Timeout

> If the device attachment is successful, we expect it to take no longer than 5 minutes.

Reviewer 3: Have a question regarding the device attach timeout, do we expect backoff if it timeout consistently?

Author: Thank you for your question regarding the device attach timeout. At this stage, backoff is one of the options we are considering if the timeout occurs consistently. However, we have not yet finalized the details. Additionally, if a timeout occurs, it might be better to consider all attempts as failures and reschedule them.

Reviewer 3: Got it, thanks

pool: Node1
driver: gpu.nvidia.com
nodeName: Node1
devices:
  - name: device3
  ...
```

During the scheduling cycle, the DRA plugin reserves a `ResourceSlice` for the `ResourceClaim`.
In the binding cycle, the reserved `ResourceSlice` is assigned during `PreBind`.
If a fabric device is selected, the scheduler waits for the device attachment during `PreBind`.
The composable controller performs the attachment operation by checking the flag of the `ResourceClaim`.
Once the attachment is complete, the controller updates the `ResourceClaim` to indicate the completion of the attachment.
The scheduler receives this update, completes the `PreBind`.

The composable controller removes device1 from the composable-device pool.
```yaml
# Composable controller publishes this pool
kind: ResourceSlice
pool: composable-device
driver: gpu.nvidia.com
nodeSelector: fabric1
devices:
  - name: device2
  ...
```

If the vendor's plugin responds to hotplug, device1 will appear in the ResourceSlice published by the vendor.
```yaml
# Vendor DRA kubelet plugin publishes this pool
kind: ResourceSlice
pool: Node1
driver: gpu.nvidia.com
nodeName: Node1
devices:
  - name: device3
  ...
  - name: device1
  ...

> ```

Reviewer 4: So the idea is to trigger the fabric-attached GPU to be hot-plugged to the node and then made visible in the "standard" ResourceSlice for consumption as a "standard" GPU. Does this mean that the resource claim actually has two devices in it? One to trigger the attachment, and one to consume the device once it is hotplugged?

Reviewer 4: Maybe it would help to see what the ResourceClaim side of this looks like.

Reviewer 5: I don't think we should be updating the resource slice. I don't see a need. We should advertise the fabric device as available from multiple nodes, with the binding gates, failure conditions, and timeout. The scheduler picks a possible node and puts that in the node name of the allocation. When the node attachment is done, the allocation already contains that node name, we already know what node we are going to - the resource slice already has the device marked as used. I see no need to update that.

Basically, I think we can do this whole thing with just those 3 new fields (binding gates, failure conditions, timeout) in the ResourceSlice Device, and the existing NodeName and Conditions in the claim status.

Author: @johnbelamaric 

> I don't think we should be updating the resource slice. I don't see a need. We should advertise the fabric device as available from multiple nodes, with the binding gates, failure conditions, and timeout. The scheduler picks a possible node and puts that in the node name of the allocation. When the node attachment is done, the allocation already contains that node name, we already know what node we are going to - the resource slice already has the device marked as used. I see no need to update that.

I understand your perspective. However, there are a few considerations that lead us to believe updating the ResourceSlice is necessary:

1. **Definition of Fabric Devices**: Fabric devices are defined by the vendor managing the fabric, not by the device vendor. Therefore, the ResourceSlice corresponding to fabric devices is output by the fabric device vendor.

2. **Handling Device State After Attachment**: After a device is attached, if the vendor's DRA driver restarts for any reason, the same device might be registered in the node-local ResourceSlice. This leads to the problem of a single device being registered in two places (the Node selector's ResourceSlice and the node-local ResourceSlice).

3. **Avoiding Duplicate Registrations**: To address this issue, we need to accurately reflect the actual state of the device in the ResourceSlice. Specifically, devices in the fabric should be registered in the Node selector's ResourceSlice, and once connected to the node, they should be treated as node-local devices.

4. **Vendor DRA Management Post-Attachment**:
   - During container initialization, the device vendor's DRA driver needs to initialize the device using NodePrepareResource(). The ResourceClaim (including the DeviceID and the corresponding ResourceSlice) passed to NodePrepareResource must be interpretable by the device vendor's DRA driver.
   - The ResourceSlice defined by the fabric device vendor may not be interpretable by the device vendor's DRA driver. For instance, in NVIDIA's implementation, the ResourceSlice is managed using internal index information that is only known after attachment, making it impossible for the fabric device vendor to define the ResourceSlice.
   - Therefore, separate management is necessary.

5. **Operational Consistency**: The CDI DRA should remove the device from the ResourceSlice after attachment, and the vendor DRA should add the attached device (requiring a rescan function). This ensures that the device is correctly managed and avoids any conflicts or duplicate registrations.

From the scheduler's perspective, even if the controller updates the ResourceSlice, the scheduler's operations remain unchanged. This is because updating the ResourceSlice would trigger the binding failure flow, leading to a reschedule. Therefore, the scheduler will handle the process consistently regardless of the ResourceSlice updates.

By updating the ResourceSlice, we ensure that the device's state is accurately reflected, preventing issues with device registration and ensuring smooth operation.

Reviewer: I've argued against moving "the same device" from one pool to another before. This leads to race conditions such that the same device is allocated for the claim which triggered attaching, is seen as a "new local device", and then gets allocated for another claim.

> The ResourceSlice defined by the fabric device vendor may not be interpretable by the device vendor's DRA driver. 

Drivers don't read ResourceSlice information. They produce it. What they consume is the ResourceClaim status, which includes the driver/pool/device tuple. We talked before about having to have some level of support in vendor DRA drivers for composable disaggregated infrastructure and a proxy which wraps it, see https://github.com/kubernetes/kubernetes/issues/124042#issuecomment-2510592082 and onwards.

Author: I understand that @pohly is opposed to moving "the same device" from one pool to another. While I would like to discuss the implementation of the controller, I want to avoid interrupting the KEP process due to this issue. Therefore, I suggest we either return to kubernetes/kubernetes#124042 or continue the discussion in a separate issue. What do you think?

Additionally, here are my current thoughts on the concerns:

### Race Condition Concerns

- For the fabric ResourceSlice, while it is allocated to the ResourceClaim, it is not used by anyone else. In the case of composable disaggregated infrastructure DRA, this ResourceSlice is always released by the scheduler, so it is never actually used.
- When the device appears in the node-local ResourceSlice, another pod might attempt to use it. However, this is acceptable. If this happens, the rescheduled pod will reallocate the fabric ResourceSlice and attach a new device, so it does not pose a real problem.
- The ideal scenario might be to use the same ResourceSlice for the same device without moving it. However, this approach encounters issues with `NodePrepareResource()`, as described below.

### Additional Considerations

Fabric devices are abstracted and may not have a specific identity until they are attached to a node. In such cases, the information that becomes known only after attachment must be communicated from the fabric vendor to the device vendor's DRA driver (via `NodePrepareResource`). However, this method of communication is not standardized. Standardizing this and implementing it in each device vendor's DRA driver to interpret the contents of the ResourceClaim would be necessary. Additionally, a mechanism to prevent the device vendor's DRA driver from publishing a ResourceSlice that includes devices from the fabric ResourceSlice would be required.

While it is possible to prevent the vendor's ResourceSlice from being published entirely, devices directly connected to PCIe slots within the server need to be published in the vendor's ResourceSlice. In this case, the vendor's DRA would need a deny list to prevent specific devices from being published. Since the device list in the fabric ResourceSlice changes, this deny list would also need to be updated in accordance with changes in the fabric ResourceSlice. While this process is time-consuming and challenging to implement, in the moving approach, the vendor DRA would only need to implement a periodic rescan functionality.

- The race condition concerns with the moving pattern can be addressed by the controller.
- In cases where the fabric device is abstracted and does not have a specific identity until attached to a node, the ResourceSlice defined as a fabric device is only informational and does not lead to a real race condition. Even if the fabric device's ResourceSlice is reused due to a delay in deletion, a well-designed controller can prevent issues.
- Even if the fabric device has a specific identity, the controller can use override functions to manage it effectively. In the case of fabric ResourceSlices, there is no specific identity information, so even if the same device ID is selected, a different device is actually attached.

Reviewer: > While I would like to discuss the implementation of the controller, I want to avoid interrupting the KEP process due to this issue.

I'm a bit reluctant to move forward with a feature when it's not clear yet how it will be used.

> Therefore, I suggest we either return to https://github.com/kubernetes/kubernetes/issues/124042 or continue the discussion in a separate issue. What do you think?

This discussion thread is more focused on one particular aspect than the issue - let's continue here.

> For the fabric ResourceSlice, while it is allocated to the ResourceClaim, it is not used by anyone else. 

Let's be a bit more precise with the wording: slices don't get allocated for a ResourceClaim, devices in the slice do and are identified with driver/pool/device names. The point of the allocation is to prevent double use of the same device.

> Once the attachment is complete, the controller requests rescheduling instead of completing the PreBind process. The controller changes a flag in the ResourceClaim to indicate that rescheduling is needed. During rescheduling, the scheduler selects the node-local device based on the updated node-local ResourceSlice published by a vendor DRA.

This seems very fragile to me. There's no guarantee that a) the scheduler has seen the updated slices by the time that it tries rescheduling and even if it has that b) it will pick the now node-local device. It might pick some other fabric device and the cycle will continue. I know you wanted some kind of "prefer node-local devices" feature. Did that make it into this KEP?

Reviewer 5: Ok, this is interesting. So, you are saying there are different drivers for the fabric and for the device? I am not clear on what the ResourceClaim would look like in that case. Would you need two, one to initiate the fabric attachment and one to allocate the device? That doesn't seem like a great UX. I wonder if there is another approach.

Could the fabric DRA driver instead worked as a sort of proxy in front of the device DRA driver? So, the vendor DRA driver is pointed to "register" with the fabric DRA driver on the node, and it publishes its ResourceSlices there. The fabric driver registers with kubelet instead of the vendor driver, and also "republishes" them to the API server, adding in the additional BindingGates, BindingFailureConditions, and BindingTimeout fields. The user making the ResourceClaim just creates a claim for the vendor device. The fact that *this particular ResourceSlice* is fabric attached is effectively transparent to the user. The scheduler picks it, and follows the PreBind dance described above; once binding is done, the scheduling goes through (constrained to the node by the NodeName in the allocation). The kubelet calls the fabric driver which calls through to the vendor driver.

Reviewer 5: I think we should give consideration to the idea of having there be some coordination or understanding between the fabric driver and the underlying vendor DRA driver. We have to put the knowledge that there are "potential" devices somewhere - I would prefer it be hidden from view of the user. I don't want the user to have to understand this at all. In current plan, what ResourceClaims does the user have to write?

Author: OK, we will continue the discussion here.

> Let's be a bit more precise with the wording

I apologize for any confusion. You are right; we need to be precise in distinguishing between ResourceSlices and the devices within them. I will be more careful with this.

> a)the scheduler has seen the updated slices by the time that it tries rescheduling

This can be managed by the controller. The controller will request rescheduling only after the node-local ResourceSlice has been updated to ensure the scheduler has the latest information.

> b) it will pick the now node-local device

We can lower the priority of fabric devices. We plan to use the prioritization feature discussed in #5034, which allows us to set priorities for different types of devices. (Will this be ready by 1.33?)

> Ok, this is interesting. So, you are saying there are different drivers for the fabric and for the device? I am not clear on what the ResourceClaim would look like in that case.

There is a controller that manages fabric devices. The devices in the ResourceSlice published by this controller will have the device vendor's driver name set. This means users can specify both fabric and node-local devices using the device vendor's driver name. Users do not need to be aware of whether the device is on the fabric, simplifying the user experience.

In our current plan, to minimize changes outside the KEP, the fabric device controller will wait for the attached device to be published by the vendor's driver after attaching the device to the node. Then, it will request rescheduling from the scheduler.

Reviewer: > The controller will request rescheduling only after the node-local ResourceSlice has been updated to ensure the scheduler has the latest information.

The controller cannot be sure that the scheduler has already received the information. It can be sure that itself has it if it watches slices, which makes it likely that the scheduler also has it, but there's still a race.

> We plan to use the prioritization feature discussed in https://github.com/kubernetes/enhancements/pull/5034, which allows us to set priorities for different types of devices. (Will this be ready by 1.33?)

No. I took it out when making device taints separate fields. Having only `kubernetes.io/priority` left as a pre-defined field seemed odd.

> In our current plan, to minimize changes outside the KEP, the fabric device controller will wait for the attached device to be published by the vendor's driver after attaching the device to the node.  Then, it will request rescheduling from the scheduler.

So what you have described so far is basically a mechanism that turns network-attached devices into local devices. It uses the attach feature in this KEP, without ever going through the "happy path" (scheduler requests attaching, waits, then proceeds with pod binding). It is a bit odd to define and implement a feature where the main use case is never using half of the feature, but let's put that aside for a moment.

How does the other direction work? Under which conditions does a local device become network-attached again? Remember that it is advertised in a slice, so at any time the scheduler could bind a pod which depends on such a device. This is either a one-way street (attach devices until all are locked to one particular node) or will most likely have race conditions that can lead to pods that are stuck on a node.

Reviewer: > Could you clarify the specific cases where you are concerned about race conditions occurring? While the controller cannot directly verify that the scheduler has the latest slice, it ensures rescheduling only after the node-local ResourceSlice is updated. This should mean the scheduler will see the updated node-local ResourceSlice.

"Should" is not the same as "must have".

The bad outcome of this race is:
- Device is attached.
- Controller sees new ResourceSlice from node driver.
- PreBind fails
- Pod gets scheduled again, without the scheduler having seen that updated ResourceSlice.

Pretty unlikely because of the delays caused by pod rescheduling, but not impossible.

> How about that sorting the pools when creating them in GatherPool(). Alternatively, will there be a way to set priorities in the future?

A scoring KEP would be the right place to add it.

> The mechanism described indeed turns network-attached devices into local devices. While it may seem odd to use the attach feature without going through the "happy path," this approach allows us to leverage existing mechanisms to manage device attachment and ensure consistency.

The point is that this mechanism doesn't exist yet. If it isn't going to be used as described ("happy path"), then perhaps there are better alternatives. For example, you could run some kind of "device autoscaler" which reacts to pods not getting scheduled because devices need to be attached first. I'm not saying that this is a better solution. If it isn't, the reasons would have to be discussed under "Alternatives" of the KEP.

> For example, if a device is no longer needed on a particular node, the controller can detach it and update the ResourceSlice to reflect its availability as a network-attached device. The controller will monitor the node-local ResourceSlice and ResourceClaim to identify devices not in use by any pod. If a device remains unused for a certain period, the controller will override it to offline status, preventing further use by pods, and then proceed with detaching the device. Using the override (or taint?) feature from https://github.com/kubernetes/enhancements/pull/5034 ensures that the device is safely detached without causing pods to get stuck.

This is still racy. You might set the taint at the same time as a pod being scheduled. It then has to be deleted and created anew. But if then the device gets removed before the pod got evicted, it will be stuck permanently because the device taint controller won't know that it needs to evict. All of that can be made less likely with delays, but that just reduces the risk. It does not make removal 100% safe.

If the device never gets removed, the ResourceSlice will keep growing, or you need some pretty elaborate scheme how the local DRA driver is supposed to handle hotplugging (create new device or reuse some existing ResourceSlice entry).

Reviewer: > In this case, the number of node-local and fabric devices remained the same, so I can't quite understand "ResourceSlice growing".

How does the DRA driver know that it should or shouldn't publish a device in its ResourceSlice? A naive approach would be to create a new entry each time a device is seen. Then the slice grows by one each time a device gets attached.

The driver can also create an entry once it sees a new device and remove it when the device is gone again. That leads to the problem described above, with the device and thus the taint being removed too soon. I think you are assuming that the driver uses this approach, but it's not spelled out anywhere.

Reviewer 6: As an assumption, Composable DRA exposes a free device on the fabric that is not yet connected to a node as a ResourceSlice.

Composable DRA refreshes the ResourceSlice periodically (every 10 seconds). This means that it reflects the latest list of devices on the fabric.
Composable DRA does not "detect attach or detach to nodes and update them immediately in event handlers, etc." This is because it is difficult for a Composable DRA running on K8s to cover all cases where a ResourceSlice needs to be updated, such as when a new device is physically added to the fabric.

We also expect vendor DRAs to periodically update the list of devices connected to the node as a ResourceSlice. This requires the rescan function to be run periodically.

> I think you are assuming that the driver uses this approach, but it's not spelled out anywhere.

Sorry for that. We will update KEP later.

Reviewer: > Composable DRA refreshes the ResourceSlice periodically (every 10 seconds). This means that it reflects the latest list of devices on the fabric.

How does it name the devices in the pool? Is each name guaranteed to be unique as it comes and goes?

This affects race conditions can fail. In the worst case, the scheduler might pick a device named "A" which is currently GPU #12345, but at the time when it comes to attaching, "A" is a different GPU with a totally different model and that one gets attached instead.

Reviewer 6: We design it to have a unique device name.
Note, however, that the device name is not an identifying name (for example, UUID).

In our Composable System, users attach devices by specifying the model name and number of devices they need.
And until the device is actually attached to the node, the user does not know which specific individual is attached.

Because of this concept, the Pool and ResourceSlice exposed by Composable DRA are separate for each model.
The devices in the Pool for each model have unique device names, but are essentially information about how many devices of this model are in the ResourceSlice.
Composable DRA also add a model name and so on into the attributes of each device.

 Therefore, the problem you are concerned about must not occur. We will also describe the above to the KEP.
 
 
 
![composable-resourceslice](https://github.com/user-attachments/assets/06ad4368-a2d0-406f-8d97-db7efe46fa93)

Reviewer 6: @pohly 
>> The point is that this mechanism doesn't exist yet. If it isn't going to be used as described ("happy path"), then perhaps there are better alternatives. For example, you could run some kind of "device autoscaler" which reacts to pods not getting scheduled because devices need to be attached first. I'm not saying that this is a better solution. If it isn't, the reasons would have to be discussed under "Alternatives" of the KEP.
>
> We will consider about Alternatives, and will update the KEP.

We have included alternatives in the KEP.
I considered "device autoscaler" which is a device version of CA.
If Pod become Unschedulable state, device autoscaler tries to attach new devices.
And it also try to detach devices if they have not been used for a period of time.
This is similar to the concept of CA.

However, if CA and device autoscaler is running independently, CA may add a node with a device at the same time as the device autoscaler attaches the device.  This is a wasted resource addition.
Therefore, we listed the idea of putting device-scale functionality in CA as an alternative.

Reviewer: True, which is a good argument for having a ResourceSlice with devices that potentially can be attached because then both scheduler and CA know that no new node is needed. Definitely put that into the KEP if it isn't there already (haven't checked).

Reviewer 5: Above I stated:

> I think we should give consideration to the idea of having there be some coordination or understanding between the fabric driver and the underlying vendor DRA driver. We have to put the knowledge that there are "potential" devices somewhere - I would prefer it be hidden from view of the user. I don't want the user to have to understand this at all. In current plan, what ResourceClaims does the user have to write?

I see that you covered part of this in the discussion above - the user writes claims using the vendor driver name. That's good. I still have concerns about the "hand off" of the claim - @pohly's point of conversion from fabric-available to node-local and back. 

Have you thought about if this can run more smoothly if the vendor driver *knows* there are fabric-attachable devices that can appear at runtime? I understand it's less desirable to have the vendor driver need to be aware of this, but I do wonder if it would simplify things. Could there be some "attachment" interface between the node-local driver and the fabric system, that avoids some things needing to be done via the K8s control plane?

Reviewer 6: I thought about it, but I think it is difficult to simplify it because it requires a big modification to vendor DRA.

First, I can think of a way for the vendor driver to recognize fabric devices:
The vendor driver recognizes the fabric device through fabric ResourceSlice. In other words, if the NodeSelector's label matches the node and the device in that ResourceSlice has the fabric attribute, the vendor driver on it's node can indirectly recognize the fabric device.

In this case, the following processing flow is possible.
1. When the scheduler assigns a fabric device to a ResourceClaim, it should selects the fabric device that the vendor DRA recognizes (or rather exposes by node-local ResourceSlice). (Fabric ResourceSlices are not exposed to users)
2. The composable DRA or vendor DRA detects this and performs the attach process. (If vendor DRA does the attach process, we need to prepare an attach IF)
3. After the attach is complete, the vendor DRA updates the node local ResourceSlice. In other words, information that can be obtained from the OS should be added to the attributes of the device.
4. After this update is complete, the vendor DRA updates the status of the device in the ResourceClaim to Attached.
5. The scheduler proceeds to the Bind process. In this case, there is no need to reschedule or convert fabric-available to node-local.

The following are concerns.
Devices in the fabric are not determined to which node they are attached until the scheduler assigns them to a ResourceClaim. Therefore, as soon as we attach to one node, the vendor drivers on other nodes need the ability to synchronize to know that the device is no longer available. This requires a mechanism to notify the drivers of each node from the fabric system, or a synchronization process to be performed by communicating between drivers of each node.
With this in mind, I think the vendor DRA will require complex and large modifications.

Reviewer 6: > True, which is a good argument for having a ResourceSlice with devices that potentially can be attached because then both scheduler and CA know that no new node is needed. Definitely put that into the KEP if it isn't there already (haven't checked).

I updated the KEP.

Reviewer 2: I have completely overlooked (in my first review pass) the part that the binding eventually fails, which makes a big difference. I will get back to it later, but let me first comment on the alternatives. The one considered at the beginning (proposal2 - pod waiting as unschedulable) required a few changes in scheduler, but the last idea with autoscaling helps in a way, that no changes in scheduler are needed.

However, be aware, that once a pod cannot be scheduled, the PostFilter phase triggers preemption attempt. It's quite heavy process that slows down scheduler, but in its principle, it may cause finding a lower priority pod and preempt it. The preempted pod will in turn may have to wait for the cluster scale up (missing device attachment). AFAIK DRA does not support preemption of devices, so it may be not an immediate concern, but at some point it will.

We could say though, that the approach with autoscaling is as bad as using CA currently. This is true and that's why I think this alternative is not favorable, but technically acceptable. I still think that we should aim for being aligned to the Dynamic Resource Allocation spirit and if we fail, we should understand why, because we may be loosing a lot.

Let me explain why:
Scheduling is split into three phases
1. trying to find a node (nominate),
2. initiate binding (assume) and
3. bind (assign).

The transition between nominate -> assume should not cause any "physical changes" to the cluster, mainly because there is a legitimate transition in the opposite direction (scheduling canceled by a Permit plugin) and scheduling starts again. However, when Permit plugin allows to continue binding, scheduler expects it's a straightforward process that should succeed (even though it might be time consuming).

There are already plugins that make use of these scheduling principles (gang-scheduling), which keeps pods in assume phase, but don't let them be bound yet. Moreover, we would like to use these principles to design the reservation feature, in which a few gang-scheduling attempts (scheduling passes) eventually produce optimized resource reservation (aka scheduling plan). Once such a plan is created, it's expected that the binding process eventually succeeds. So the critical part is that reaching the assume phase should not cause any physical action (side effects), as it may be repeated several times.

So the alternative of going through the autoscaling process is not completely bad (does not break theses assumptions), but the funny thing is that the autoscaling process could be in fact implemented as the dynamic allocation, where nodes are offered as allocatable devices. So the approach described above would allow to make the scaling process a part scheduling process (a part of such execution plan), allowing scheduler try to place pods much more optimally (in a few attempts) and much faster (without waiting for the autoscaling turnarounds).

A few other comments:
> This can be managed by the controller. The controller will request rescheduling only after the node-local ResourceSlice has been updated to ensure the scheduler has the latest information.

Note that the failed binding puts node into backoff queue, which adds additional penalty time. We have the [KEP-5142](https://github.com/kubernetes/enhancements/pull/5144) proposed to reduce the backoff time to the very minimum (eliminate penalty effect), but it does not cover failed binding cases - to preserve the rate limiting effect in failure scenarios (for instance to not overload api-server).

Another bad effect of failing the binding process, is that other resources which were reserved for the planned pod may be taken by other pod in the consecutive cycles and eliminate this node (where the device got attached to) from the "final" scheduling cycle. The "autoscaling" approach also has exactly the same inconvenience (other resources may be taken), which would lead to a long recovery process which eliminates the device from being used for that time being. So another dirty hack to consider is to make the attachment process a part of the preemption process (PostFilter), as in this process the node resources gets reserved until preemption happens (device attachment in our case).

I may still not understand everything in DRA, so not sure if this is possible, but I'd try to solve the problem by having proxy devices (as John suggested). Any attachable device would be offered as the final device though proxy (to avoid the problem of late (lazy) ResourceSlice creation), but available with many nodes. Scheduler could consider it while analyzing different reservation options (execution plans), but allocate it at the time of reservation (plan) execution - binding. Without such proxies, scheduler would not be able to analyze several execution plans without going through autoscaling process in the meantime, so in unreasonably long time.

This approach would be especially useful if the attached devices were partitionable devices. Assuming that scheduled pods need just "a single partition", the proposed autoscaling approach could in fact cause attaching full devices per a pod in need. On the opposite side, the approach with proxies would allow to allocate those single partitions and cause binding as little big devices as possible.

Sorry for such a long comment. I hope it won't be confusing, especially since it describe non-existent functionality of reservations and execution plans.

Reviewer 5: Thanks @dom4ha. Interesting discussion. A few points/questions:
> So another dirty hack to consider is to make the attachment process a part of the preemption process (PostFilter), as in this process the node resources gets reserved until preemption happens (device attachment in our case)

So, you are saying do this in PostFilter rather than PreBind, because the way the scheduler will respond to a failed attachment in PostFilter is better than in PreBind?

>This approach would be especially useful if the attached devices were partitionable devices. Assuming that scheduled pods need just "a single partition", the proposed autoscaling approach could in fact cause attaching full devices per a pod in need. On the opposite side, the approach with proxies would allow to allocate those single partitions and cause binding as little big devices as possible.

This is an interesting point. So, IIUC, you are saying - "What if someone is asking for a small device/partition? we should be able to attach a large device to a node, then give them the small device via the partition, leaving the rest of the larger device available for other pods.". That's definitely a possibility, but it probably doesn't work that well with the separate "fabric" device. A proxy option would work better in that case. But in order for a proxy to work, we need the vendor driver to know such devices are "available" via an attachment process. Which has some of the problems @hase1128 laid out (in particular, all possible nodes knowing when the devices are no longer available). I wonder if we can figure out a way for drivers to advertise "potential" devices (cc @sunya-ch, as this looks like "provisionable"...). I think this is something we can explore more in the future.

Reviewer 2: > So, you are saying do this in PostFilter rather than PreBind, because the way the scheduler will respond to a failed attachment in PostFilter is better than in PreBind?
>
I think that Aldo had even a better suggestion to use Permit extension point instead. The most important thing to achieve are:
1. keep resources reserved ("assumed" in Permit or "nominated node" in PostFilter or PreBind)
2. avoid reporting a failure (binding failure) after device attachment (to avoid penalty)
3. avoid triggering preemption process, as attachable devices should be treated as available resources
4. trigger final scheduling cycle when scheduler learns about update ResourceSlice, which is possible when it waits in the unschedulable (PostFilter approach) or even better in the waiting queue (Permit approach).
 
> A proxy option would work better in that case. But in order for a proxy to work, we need the vendor driver to know such devices are "available" via an attachment process. Which has some of the problems Jin Hase laid out (in particular, all possible nodes knowing when the devices are no longer available). I wonder if we can figure out a way for drivers to advertise "potential" devices (cc Sunyanan Choochotkaew, as this looks like "provisionable"...). I think this is something we can explore more in the future.

Maybe I'm missing something, but a device should remain attached as long as there are proxy devices referring to it (they would work as a lock). If such a device needs to go for maintenance earlier, operator could request admin allocation, which could cause putting NoSchedule taint on it and wait until all pods stop using proxy devices and release all locks.

The main advantage of having proxy devices is to enable attachment at binding time, but leave scheduling purely logical analysis. This would allow scheduler to try a few different allocations while searching for the best placement (best utilization or achieving other goal), without causing any attachments. Once scheduler decides to bind a group of pods according to such optimal execution plan, proxy devices will cause necessary attachments and/or other arrangements.

Since we don't have capability to make reservation for group of pods in scheduler, we can live without proxy devices and use Permit extension point (as discussed above) as the best what we can do. In that case attachment process would be a part of scheduling process (instead of binding).

     - Once the preparation is complete, they set the conditions to `true`.

   - **Completion of the PreBind Phase**:
     - Once all conditions are met, the `PreBind` phase is completed, and the scheduler proceeds to the next step.

### User Stories (Optional)

<!--
Detail the things that people will be able to do if this KEP is implemented.
Include as much detail as possible so that people can understand the "how" of
the system. The goal here is to make this feel real for users without getting
bogged down.
-->

#### Story 1

#### Story 2

### Notes/Constraints/Caveats (Optional)

<!--
What are the caveats to the proposal?
What are some important details that didn't come across above?
Go in to as much detail as necessary here.
This might be a good place to talk about core concepts and how they relate.
-->

### Risks and Mitigations

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.

> -->

Author: Regarding collaboration with the Cluster Autoscaler, as mentioned [in this comment](https://github.com/kubernetes/enhancements/pull/5012#discussion_r1937354315), we believe that using node nomination can address the issue. Would it be helpful to explicitly state that this approach can mitigate the problem? 
If there are other issues that I might not be aware of, I would appreciate your guidance.

> -->

Reviewer 7: Got it, but I mean there's no mention on KEP (unless I missed it?) and you should explain it somewhere.

> -->

Author: OK, there's no explanation so I'll add it.

> -->

Reviewer 7: /cc @macsko @dom4ha @pohly 

Another point here is whether we regard this feature as depending on the nominated node name change or not. 
Do you know (or can you have a guess) how long the binding would take on average? If that's a few mins, then most likely the user would hit https://github.com/kubernetes/kubernetes/issues/125491.

> -->

Reviewer 7: The same thing can be said to the in-flight events problem too. That'd be a smaller change than the nominated node name stuff though.
We should clarify when the issues should be solved by, not just saying "this problem will be solved by that way, so no worries".

> -->

Author: > Do you know (or can you have a guess) how long the binding would take on average? If that's a few mins, then most likely the user would hit the issue very likely, and the criticalness of this issue is higher..

Based on our current understanding, the average binding time for devices is typically less than a minute, and we expect the time spent at PreBind to be around 2-3 minutes.

> We should clarify when the issues should be solved by, not just saying 'this problem will be solved by this GH issue, so no worries'.

I understand that we must provide a clear timeline for when these issues should be resolved. How about including both issues as criteria for the beta release?

> -->

Reviewer 7: > How about including both issues as criteria for the beta release?

Yeah, targeting beta looks fine to me.

> -->

Reviewer 2: The approach with autoscaler (as proposed by Patric in [this comment](https://github.com/kubernetes/enhancements/pull/5012#discussion_r1944540598)) bypass the issue, as there is no long binding anymore. It unfortunately inherits all CA problems, so I think a bit better hack would be to make the device attachment a part of the preemption process (PostFilter), which involves nominating node (reserving other resources). I'm not certain though how feasible such a hack is.

Regarding the issue with the volume plugin that you mentioned, I was about to write a KEP proposing how to fix it, but actually might be even better, because the reservations feature that I was talking about in that comment, could fix the issue from a completely different angle. Reserved resources for the the pod-to-bind would be visible to autoscaler (be reserved in autoscaler as well) because a reservation means putting taints on reserved resources. So long term, it would work even better than repurposing of `NominatedNodeName` as reservations could reserve also resources for not-yet-defined-pods and autoscaler would still see such reservations.

Let's take a bit less abstract example. If we have a reservation that accommodates a pod vertical scaling growth, autoscaler would take it into consideration while computing available space. It could not be achieved though using `NominatedNodeName`, which would still refer to the old pod (prior scaling).

I know we are far from reaching the point we have reservations implemented (or even designed), so we need to discuss what is our priority to fix this issue.

> -->

Author: Thank you for your proposed solution. We have set a milestone to resolve this issue in the beta phase, so let's continue our discussions with that goal in mind.

> -->

Reviewer 2: I understand that, however considering the design with abandoning binding, it would be wiser to use PostFilter phase from the very beginning (instead of PreBind), since you could initiate attaching and set nominated node there. It would solve many problems mentioned in the review (scheduler restarts, Cluster Autoscaler pod visibility, cache growth and finally stealing resources by other pods, which could lead to leaving the device stuck unused). Implementation wise there is no big difference, but operationally it there is.

> -->

Reviewer 2: To be super clear here, we planned to set nominated node name in PreBind as well to address the CA issue that Kensei brought up here, so in fact it would also address the issue of CA visibility, scheduler restarts and the risk of stealing resources.

@sanposhiho, @Huang-Wei, @macsko what would you recommend here? Using PreBind has one advantage of doing things async by design (binding is async)

> -->

Reviewer 6: As you point out, if we wait in PostFilter, scheduler cannot schedule the next Pod until the attachment is complete.

Also, in the PostFilter phase, the device in ResourceSlice has not yet been assigned to the ResourceClaim, so the Composable DRA Controller cannot determine at this point whether attachment is necessary.

> -->

Reviewer 2: Sorry I haven't explained it clearly. In PostFilter the plugin would request the device attachment, set `NominatedNodeName` and return with "unschedulable" status. Pod would wait in uschedulable queue until the attachment is successful and perform scheduling once again. Having a node nominated help us stick the pod to the initially selected node.

The only downside is that PostFilter is usually used for preemption, but I guess it's not set in a stone. The second downside is that PostFilter is synchronous, so performing api calls are slowing down scheduling. We have however may have solutions to that problem as well.

What we avoid is keeping a pod in assumed state for a longer period of time and then release it as a "binding failure" without guarantees that the next scheduling cycle would see the updated ResourceSlice or/and waiting in backoff queue in as penalty for the failed binding.

> -->

Reviewer 6: I see, I understand your thought and consideration.

>The only downside is that PostFilter is usually used for preemption, but I guess it's not set in a stone. The second downside is that PostFilter is synchronous, so performing api calls are slowing down scheduling. We have however may have solutions to that problem as well.

I want to hear comments from other SIG-Scheduler guys whether there is good solution.

>What we avoid is keeping a pod in assumed state for a longer period of time and then release it as a "binding failure" without guarantees that the next scheduling cycle would see the updated ResourceSlice or/and waiting in backoff queue in as penalty for the failed binding.

Is there any way to mitigate this penalty? Is this reasonable to consider that we use PostFilter rather than PreBind?

> -->

Reviewer 8: PreBind is somewhat busy-waiting: there will be an active goroutine waiting for the process to complete. This can lead to goroutines leakage if there is a problem attaching the fabric, plus memory leak due to inflight pods.

Not sure if PostFilter is the right place, as it only runs if the Filters didn't pass. A more appropriate place could be Permit, which happens after we "assume" the pod, so the resources are reserved in-memory.
But this is no longer busy waiting. The pod goes back to the unschedulable stage (IIRC), waiting for an event (in this case, the fabric finished attaching) to make it schedulable again.

> -->

Reviewer 7: There're couple of points that have to be considered if we implement it with PostFilter:
- Difference between nominatedNodeName vs assumed node: If we go with PostFilter, the node is reserved for the DRA pod with nominatedNodeName while binding devices, while if PreBind, the reservation is made with "assume", which is internal node reservation that the scheduler has. They're similar, but one big difference is how to treat higher priority pods; if pod-a is nominated on node-a, and a higher priority pod comes to the scheduling cycle next, the pod can be scheduled on node-a, potentially taking a place of pod-a (i.e., if the cluster is super packed). OTOH, if pod-a is "assume"ed on node-a, no one can take the place of pod-a, even a higher priority pod cannot.
- Need to distinguish DRA pods normally failed vs DRA pods waiting for binding: So, if we go with PostFilter, there're DRA pods that are failed at Filter (normal failures) and DRA pods that are waiting for devices to be bound. We don't have any way to deliver "why this pod is failed by DRA plugin" to QHint. We need to somehow distinguish those two.
- How to unbind: let's say the device binding somehow doesn't go well. With PreBind idea, we have timeout; once we reach the timeout, we stop waiting, and perform necessary unbinding operation. We need to think how we can implement this timeout and consequence unbinding with them.

> -->

Reviewer 7: Unless we can get _simple_ answers for the above points, I think the current design is still better, especially given we'll, either way, work on having nominated node name for pods at binding cycles.

> -->

Reviewer 2: Permit sounds even better, as it's after the scoring, so the best placement can be determined with regular scheduling mechanism (instead of just by a single plugin in PostFilter).

IIUC Permit needs to return "wait" instead of deny to not release allocated resource, is that correct?

> -->

Reviewer 6: @alculquicondor 
CC: @dom4ha @macsko  @sanposhiho @pohly @johnbelamaric @klueska 

I would like you to approve this KEP by the enhancement freeze of K8s1.33(Feb. 13th). I think we have decided on most of what is needed for alpha.

As I understand it, the only main discussion point left for alpha is whether to use Permit or PreBind.

For example, how about using the current design (PreBind) for alpha and continuing the discussion here on whether to use Permit or other better way for beta?

> -->

Reviewer 7: > A more appropriate place could be Permit, which happens after we "assume" the pod, so the resources are reserved in-memory.
But this is no longer busy waiting. The pod goes back to the unschedulable stage (IIRC), waiting for an event (in this case, the fabric finished attaching) to make it schedulable again.

We assume the pod before Permit, but we unassume (=`sched.cache.ForgetPod`) the pod if Permit fails. So, this wouldn't work. As far as I remember, we don't have any way to assume pods, but move them back into the queue. 

> IIUC Permit needs to return "wait" instead of deny to not release allocated resource, is that correct?

If it returns wait, the pod goes to the binding cycle and then wait for the signal to permit the pod to proceed to the rest of the binding cycle. So, it'd be essentially the same as using PreBind (i.e., waiting within a goroutine)

> -->

Reviewer 7: This discussion is about the fundamental part of the implementation, and I'm not sure if _"we have decided on most of what is needed for alpha"_ without this discussion's conclusion.

> -->

Reviewer 6: Understood. However, based on your comment about Permit, I think it would be better to use PreBind, even though there are penalty issues.

> -->

Reviewer 7: Yeah, so I don't see any other feasible ideas apart from PreBind, based on my understanding; using Permit is invalid and using PostFilter has to address the points I raised above. So, actually, I'm +1 for the current design. (ofc I'm just one of the reviewers, even not approvers, and you need agreement from others too though)

> -->

Reviewer 7: @hase1128 @KobayashiD27 
FYI: Luckily we have a sig-scheduling meeting from 6PM JST today. @alculquicondor won't be there, but @macsko, @dom4ha, and I will be. You can attend it for a prompt discussion, if you want.

> -->

Reviewer 6: This is good information. I am not good at English but I will participate.
(@dom4ha has already listed several KEPs including this KEP in today's agenda.)

> -->

Reviewer 7: Nice. No worries, neither am I.

> -->

Reviewer 7: The brief conclusion from the meeting: we can do it with PreBind _for now_ because looks like PostFilter or Permit (with the current form) don't satisfy the requirement. But, there's a couple of concerns that raised like how we should reschedule the pods that faced the failure of device bindings etc. We should continue the discussion and consider solutions for them, including some better long-term ideas.

> -->

Reviewer 7: @hase1128 @dom4ha @macsko you can add more things from the meeting, if you want to mention.

> -->

Reviewer 6: Thank you. We need to discuss this for beta release.
I hope @alculquicondor approves the KEP as alpha.

> -->

Reviewer 6: This is a case-by-case basis.
The following are cases where we may want to try a different node
- If the node is running out of resources due to other Pods during PreBind
- If some physical problem caused the attachment to the node to fail

Ideally, when the Composable DRA Controller store the result of an attachment failure in the ResourceClaim (or more precisely, sets the conditions of BindingFailureGates to true), it would be good to have a mechanism that allows DRA driver to specify whether to select the same node or a different node depending on the reason for the failure.

> 

Author: We believe that the scheduler's restart should not pose an issue, as the decision to wait is based on the Conditions of the ResourceClaim. After a scheduler restart, if the device attachment is not yet complete, the scheduler will wait again at PreBind. If the attachment is complete, it will pass through PreBind. 
I will add this information to the KEP.

> 

Reviewer 7: >  if the device attachment is not yet complete, the scheduler will wait again at PreBind.

I don't see how to do this in KEP. How would you ensure the scheduler pick up the same node for the Pod after the restart?

> 

Reviewer 7: Note that the scheduler after a restart just re-schedule all pods that were in binding cycles but had not yet bound before the restart. It doesn't know where those pods were meant to be bound at the previous scheduler.

> 

Author: I think it is acceptable for the scheduler to select a different node after a restart. If a different node is selected, another device will be attached on that node. 
If a device was in the process of being attached before the scheduler was restarted, the controller will detach it after a certain period if it determines that the device is no longer needed. (This is just a plan)

> 

Reviewer 7: But, it's obviously better (efficient) to select the same node that the pod was about to be bound before the restart, right? Can't we explore that way? like selecting the same node at Filter(), for example.

> 

Author: I completely agree that it would be more efficient to select the same node that the pod was about to be bound to before the restart. Unfortunately, with my current knowledge, I haven't been able to come up with a reliable method to handle this scenario effectively.
I would greatly appreciate any ideas or references you might have on this topic. Exploring the possibility of selecting the same node at the Filter() stage sounds promising, and I'm open to further discussion on how we can implement this.
(However, I'm concerned that aiming for an alpha release in v1.33 might not give us enough time to address this issue thoroughly.)

> 

Author: Regarding the suggestion to schedule the pod on the same node, we need to discuss the following scenarios:

- In case of a failure, we might want to try a different node.
- During rescheduling, if another pod is deployed on that node and uses the resources, the rescheduled pod might not be able to be deployed. Therefore, we need logic to prioritize the rescheduled pod on that node.

These discussions will require time, so we would like to address this after the alpha. Additionally, since we do not consider this a mandatory requirement, we want to make it clear that this will be handled in the later release.

> 

Reviewer: Where is the information to prefer one node for a pod stored? The scheduler plugin is stateless across pod scheduling attempts.

> 

Reviewer 7: I'm not sure, just spreading the ideas. But, rethinking about it, I just realized that, once we've done [the nominated node stuff](https://github.com/kubernetes/enhancements/pull/5012#discussion_r1945864184), 
1. pods that were in binding cycles but had not yet bound before the restart should have nominated node names
2. the scheduler is restarted (the nominated node names are stored in pod objects and obviously won't lost)
3. the scheduler after restart try the nominated node first and then try other nodes if the nominated node isn't available.

So, the nominated node improvement solves this point together.

> 

Reviewer 6: >Back to this point; what do you mean by "this is just a plan"? What's the current status of it? like, is it designed (like with another KEP), but not yet implemented?

This is the design of the Composable DRA controller that we implement. I assume he used the word "plan" because the implementation is not yet complete. However, we are going to implement controller based on the above design.

> 

Reviewer 7: What's the current behavior without that planned feature? i.e., how would the current controller behave if this restarting scenario happens? Depending on that, we may have to track that feature as a beta blocker as well.

> 

Reviewer 6: The composable DRA controller is still under development and is not publicly available. We plan to include these features in the first release of the controller.

If the above function is not available, the process flow is as follows.
If the pod is assigned to another node after the scheduler restarts, additional device is attached to that node.

If the device attached to the original node is not used, the user can manually detach the device. (Of course, we can leave it attached to that node for future use by the Pod.)

> 

Reviewer 7: Naruhodo, now all clear to me. So, even without that feature, it's just that devices are unnecessarily left on the nodes, which is waste but is not something like crashing things up. So like surely good to have, but not a critical part.

I think then we can just track that feature as a beta feature, but not a hard requirement for alpha. (which I guess should be fine since it's already on the plan of the first release?)
Please update the KEP like that if you agree.

> 

Reviewer 6: Sure, I agreed and updated the KEP.

> 

Reviewer 2: Similarly here, picking the approach with the attachment process to be a part of the preemption process (if feasible) would allow leaving nominated node name that reserve this node resources. It would allow restarted scheduler to keep waiting for the device-in-allocation.

The approach with cluster autoscaler won't help with scheduler restart, as there would be no nominated node. It in turn can cause a device to get allocated, but due to "stollen" resources, pods-in-need cannot use that device anymore.

![proposal](proposal.jpg)

#### Device Attribute Additions

To indicate whether a device is a fabric device, an attribute is added to the `Basic` within `Device`.
This attribute will be used by the controller that exposes the `ResourceSlice` to notify whether the device is a fabric device.
To avoid impacting existing DRA functionality, the default value of this attribute is set to `false`.

```go
// Device represents one individual hardware instance that can be selected based
// on its attributes. Besides the name, exactly one field must be set.
type Device struct {
    // Name is unique identifier among all devices managed by
    // the driver in the pool. It must be a DNS label.
    //
    // +required
    Name string

    // Basic defines one device instance.
    //
    // +optional
    // +oneOf=deviceType
    Basic *BasicDevice
}

// BasicDevice represents a basic device instance.
type BasicDevice struct {
    // Attributes contains additional attributes of the device.
    //
    // +optional
    Attributes map[string]string
}
```

To indicate a fabric device, the following attribute will be added:

```yaml
attributes:
  kubernetes.io/needs-attaching:

>     bool: "true"

Author: https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/resource/types.go#L213
`Attributes` is  `map[QualifiedName]DeviceAttribute`.
The way to set `DeviceAttribute` seems to be in the format `ValueType:Value`.

>     bool: "true"

Reviewer 7: `BasicDevice.Attributes` above says it's `map[string]string`. Are they two different things?

>     bool: "true"

Author: I may be misunderstanding your concerns, but I think it's roughly correct that DeviceAttribute is actually a map[string]string. However, to be precise, I think there is only one key/value pair in this map.

>     bool: "true"

Reviewer 7: Let me re-explain. So, here on KEP, you're saying:

```yaml
attributes:
  kubernetes.io/needs-attaching:
    bool: "true"
```

This is `map[string]map[string]string`, which the first key is `kubernetes.io/needs-attaching` and the second key is `bool`.

But, at L323 on KEP, you said `basicDevice.Attributes` is `map[string]string`.

And, [the internal API you commented](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/resource/types.go#L213) has also yet another structure, which would be like the following in Yaml:

```yaml
attributes:
  kubernetes.io/needs-attaching:
    boolValue: "true"
```

So, there're three different formats in KEP and the internal API, and I'm confused which is correct or all three of them are different stuff.

>     bool: "true"

Author: Thanks. I didn't realize that there were multiple formats mixed in KEP.
Since we are going to use the existing API, the last yaml format is correct. I will fix it.
Sorry for the confusion.

type AllocatedDeviceStatus struct {
...
    // NodeName contains the name of the node where the device needs to be attached.
    //
    // +optional
    NodeName string 
}

const(
  DRADeviceNeedAttachType = "kubernetes.io/needs-attaching"
  DRADeviceIsAttachType = "kubernetes.io/is-attached"
  DRADeviceAttachFailType = "kubernetes.io/attach-failed"
)
```

#### Scheduler DRA plugin Additions
When `kubernetes.io/needs-attaching: true` is set, the scheduler DRA plugin is expected to do the following:

1. Set `AllocatedDeviceStatus.NodeName`.
2. Add an `AllocatedDeviceStatus` with a condition of `Type: kubernetes.io/needs-attaching` and `Status: True`.
3. Wait for a condition with `Type: kubernetes.io/is-attached` and `Status: True` in `PreBind` before proceeding.
4. Give up when seeing a condition with `Type: kubernetes.io/attach-failed` and `Status: True`.

Note: There is a concern that the in-flight events cache may grow too large when waiting in PreBind.
This issue will be addressed separately as outlined in kubernetes/kubernetes#129967.

#### PreBind Phase Timeout

If the device attachment is successful, we expect it to take no longer than 5 minutes.

> However, to account for potential update lags, we would like to set a fixed timeout for the scheduler to 10 minutes.

Author: Even when considering communication lags and other factors, '10 minutes' represents an exceptionally poor case among successful attachments. Based on our current understanding, we believe that a fixed timeout value should be enough.

Composable system is composed of resource pool and Composable Manager software. In Resource pool all components are connected to PCIe or CXL switches. Composable Manager controls the switches so as to create composed baremetals by software definition. It has Composable API and Operator or Kubernetes may call the API. Once composed baremetals are created user can install any operating system or container infrastructure.

This flexibility extends further with the use of fabric devices. Fabric devices can be used by multiple Kubernetes clusters, not just a single one. Each cluster expose the device as a ResourceSlice, allowing for efficient utilization of the device.

In this scenario, the ResourceSlice representing a same fabric device might be selected in multiple Kubernetes clusters simultaneously. If the attachment fails in one cluster, the pod will remain in a failed state in kubelet.

> By having the scheduler wait for the fabric device to be attached, we can reschedule the pod if the attachment fails. This approach is superior because it avoids unnecessary waiting and allows for immediate rescheduling.

Reviewer 9: I am new to this area of CDI, but would appreciate if we can show how this para is more tightly related to the proposal ? For example does this make it easier to attach requested resource directly to the node where the pod is about to be scheduled or some other example like https://cdi.liqid.com/hubfs/Inference-aaS-WP_FNL.pdf

Author: In the current implementation, device attachment is executed by NodePrepareResource() after the Pod has been scheduled. However, if the device cannot be attached to the node at this stage, NodePrepareResource() will continue to fail, and the user will see the ContainerCreating state persist (internally, it will keep failing indefinitely). This issue is particularly problematic when fabric devices are shared across multiple Kubernetes clusters, as these clusters may compete for resources, leading to situations where device attachment fails.

Implementing this proposal will help avoid the persistent ContainerCreating state. By streamlining the process of attaching the requested resource directly to the node where the Pod is scheduled, the risk of device attachment failure is reduced. Consequently, this enhances the overall reliability of the system and prevents resource wastage.

-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

<!--
This section is incredibly important for producing high-quality, user-focused
documentation such as release notes or a development roadmap. It should be
possible to collect this information before implementation begins, in order to
avoid requiring implementors to split their attention between writing release
notes and implementing the feature itself. KEP editors and SIG Docs
should help to ensure that the tone and content of the `Summary` section is
useful for a wide audience.

A good summary is probably at least a paragraph in length.

Both in this section and below, follow the guidelines of the [documentation
style guide]. In particular, wrap lines to a reasonable length, to make it
easier for reviewers to cite specific portions, and to minimize diff churn on
updates.

[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
-->

> To achieve efficient management of fabric devices, we propose adding the following features to the Kubernetes scheduler's DRA plugin. Fabric devices are those that are not directly connected to the server and require attachment to the server for use.

Reviewer: Before we do a full review, can you line wrap all paragraphs? In future updates, change as few lines as possible.

This helps with anchoring review threads and avoids false "obsoletion" by GitHub when something unrelated gets updated.

- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

To achieve efficient management of fabric devices, we propose adding the following features to the Kubernetes scheduler's DRA plugin.
A fabric device is a device that is placed outside the server via a PCIe or CXL switch.
In order to use this device from K8s, we need to control the switch to connect a specific device to a specific physical server.

In the current DRA implementation, fabric devices are attached after the scheduling decision, which leads to the following issue:

Fabric devices may be contested by other clusters.

> In scenarios where attachment occurs after scheduling, there is a risk that the resource cannot be attached at the time of attachment, causing the container to remain in the "Container Creating" state.

Reviewer 2: Can you elaborate more on the current state?
- I conclude the attachable devices currently work, so how they binds since there are no local devices before a physical device gets attached?
- Why does the container gets stuck (in "Container Creating") instead of fail binding?
- Why it's not sufficient to report a failure and retry scheduling?

Reviewer 6: First, let's explain the premise.

Fabric devices
- Fabric devices are devices that exist outside the server and are connected to PCIe or CXL switches.
- These fabric devices can connect specific devices to specific nodes by controlling the switches.
- In a large-scale fabric configuration, multiple K8s cluster nodes exist in the same fabric configuration.

When exposing such devices as ResourceSlices, the Composable DRA Controller exposes them as non-node-local (basically using node selector) devices.

In the current scheduler implementation, when a device is assigned to a ResourceClaim, it proceeds directly to Bind and the Pod is scheduled to the node. However, since the device is not attached, the Pod stops in the container creation status. If the Composable DRA Controller detects such a state and attaches it, the Pod may be created successfully. However, this method has the following problems.

ResourceSlices in each k8s cluster may expose the same device. The reason is that If we restricts which devices can be connected to which clusters in composable system, making it difficult to use devices efficiently across clusters.

If multiple clusters compete for the same device, the cluster that fails to obtain the device will be stuck in the container creation state. The Composable DRA controller has already attached the target device to a node in another cluster, so if there are no other devices available, the container will never be created.

To prevent this from happening, we proposed that the scheduler wait until the device is reliably attached to a node in the cluster in the PreBind phase. With this method, even if the device is taken by another cluster, the Bind will simply fail, preventing the creation of unnecessary containers.

Reviewer 7: Can't the controller mark the pod as failed in such cases? Then the pod is recreated by a high layer resource managing the pod (replicaset, job, etc), and a new pod is scheduled to another node.

Reviewer: > Can't the controller mark the pod as failed in such cases? 

No, not at the moment. That would be https://github.com/kubernetes/kubernetes/issues/125542.

> Then the pod is recreated by a high layer resource managing the pod (replicaset, job, etc), and a new pod is scheduled to another node.

That's a rather heavy-handed way of rescheduling a pod, with even longer latencies than putting a pod into the backoff queue.

Reviewer 7: Then, if we've got https://github.com/kubernetes/kubernetes/issues/125542, then pods get trapped into the binding failures, at least, saved with my scenario.
So, the point is that, is it a fairly common scenario that happens with a certain frequency, not just an unlucky thing that happens rarely?

Reviewer 6: Normally, if there are enough fabric devices, this problem does not occur because the Composable system can dynamically find free devices and attach them to nodes.

If the number of free fabric devices becomes low, the possibility of such problems increases, so users should request devices with caution.

Reviewer 2: Do you mean that attachable devices were already supported, but once they are shared between clusters, binding them becomes problematic?

It seems to me that recreation of ResourceSlice to a node-local device is one problem (that it requires additional rescheduling cycle), but even if we solved that problem, the another one is inability to reserve devices over all clusters they are offered, so you may never be sure whether the binding will be successful without actually trying to attach them. So it's like having a pool that you never know how big it currently is, is that correct?

Reviewer 6: > Do you mean that attachable devices were already supported, but once they are shared between clusters, binding them becomes problematic?

Yes, if we configure the fabric devices is available in multiple clusters in composable system, this problem occurs.

> So it's like having a pool that you never know how big it currently is, is that correct?
For example, if you have a Pool with 10 available devices, that actually means that a total of 10 devices are available across multiple clusters.
However, from within a K8s cluster, the status of other clusters cannot be seen, so it appears as if 10 devices are available within your own cluster.

Reviewer 2: The biggest problems for the long term support of this feature in scheduler are:
1. the inability to reserve devices (without performing an attachment) over multiple clusters (in scheduler's Reserve phase)
2. the need of rescheduling after attachment.

We don't block the development for alpha though.

the system. The goal here is to make this feel real for users without getting
bogged down.
-->

#### Story 1

#### Story 2

### Notes/Constraints/Caveats (Optional)

<!--
What are the caveats to the proposal?
What are some important details that didn't come across above?
Go in to as much detail as necessary here.
This might be a good place to talk about core concepts and how they relate.
-->

### Risks and Mitigations

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

**What if the scheduler restarts while the DRA plugin is waiting for the device(s) to be bound?**

The scheduler's restart should not pose an issue, as the decision to wait is based on the Conditions of the ResourceClaim.
After a scheduler restart, if the device attachment is not yet complete, the scheduler will wait again at PreBind.
If the attachment is complete, it will pass through PreBind.

> **Scheduler does not guarantee to pick up the same node for the Pod after the restart**

Reviewer 2: ```suggestion
**Scheduler does not guarantee to pick up the same node for the Pod after the reschedule (after binding failure)**
```
This should be either fixed on the scheduler side (nominated node added) or a need for rescheduling should be eliminated before beta.

Currently rescheduling does not guarantee picking the same node in the same way like in scheduler restart scenario.

    // has to be limited to exactly the node that was chosen when allocating the claim.
    //
    // +optional
    UsageRestrictedToNode bool

    // BindingTimeoutSeconds indicates the prepare timeout period.
    // If the timeout period is exceeded, the scheduler clears the allocation in the ResourceClaim and reschedules the Pod.
    //
    // +optional
    BindingTimeoutSeconds *metav1.Duration
}

const (
    BindingConditionsMaxSize    = 4
    BindingFailureConditionsMaxSize = 4
)
```

#### AllocatedDeviceStatus.Conditions Enhancements

The `BindingConditions` and `BindingFailureConditions` fields within `AllocatedDeviceStatus.Conditions` are used to indicate the status of the device attachment.
These fields will contain a list of conditions, each representing a specific state or event related to the device.

For this feature, following fields are added:

```go
// AllocatedDeviceStatus.Conditions contains the status of an allocated device, if the
// driver chooses to report it. This may include driver-specific information.
type AllocatedDeviceStatus.Conditions struct {
    ...
    // BindingConditions defines the conditions for binding.
    //
    // +optional

>     BindingConditions map[string]bool

Reviewer 5: We lost @pohly's previous comments here. This is here to capture the "as-was" values for these fields - the values as they were set in the ResourceSlice at the time of allocation. I think this should be `[]string`, and we need to copy over the timeout as well, and maybe the UsageRestrictedToNode?

Therefore, it publishes a list of free fabric devices as `ResourceSlices`.

The structure we are considering is as follows:

```yaml
# Composable controller publishes this pool
kind: ResourceSlice
pool: composable-device
driver: gpu.nvidia.com
nodeSelector: fabric1
devices:
  - name: device1
    UsageRestrictedToNode: true
    ...
  - name: device2
    UsageRestrictedToNode: true
    ...
```

The vendor's DRA kubelet plugin will also publish the devices managed by the vendor as `ResourceSlices`.

```yaml
# Vendor DRA kubelet plugin publishes this pool
kind: ResourceSlice
pool: Node1
driver: gpu.nvidia.com
nodeName: Node1
devices:
  - name: device3
  ...
```

During the scheduling cycle, the DRA plugin reserves a `ResourceSlice` for the `ResourceClaim`.
In the binding cycle, the reserved `ResourceSlice` is assigned during `PreBind`.
If a fabric device is selected, the scheduler waits for the device attachment during `PreBind`.
The composable controller performs the attachment operation by checking the flag of the `ResourceClaim`.
Once the attachment is complete, the controller updates the `ResourceClaim` to indicate the completion of the attachment.

> The scheduler receives this update, completes the `PreBind`.

Reviewer 2: ```suggestion
The scheduler receives this update, completes the `PreBind`.
Note that in Composable system, it abandons binding by reporting a failure, forcing scheduler to rerun scheduling cycle and bind to the attached device, which is visible now as a local device.
```

![proposal](proposal.jpg)

#### Device Attribute Additions

To indicate whether a device is a fabric device, an attribute is added to the `Basic` within `Device`.
This attribute will be used by the controller that exposes the `ResourceSlice` to notify whether the device is a fabric device.
To avoid impacting existing DRA functionality, the default value of this attribute is set to `false`.

```go
// Device represents one individual hardware instance that can be selected based
// on its attributes. Besides the name, exactly one field must be set.
type Device struct {
    // Name is unique identifier among all devices managed by
    // the driver in the pool. It must be a DNS label.
    //
    // +required
    Name string

    // Basic defines one device instance.
    //
    // +optional
    // +oneOf=deviceType
    Basic *BasicDevice
}

// BasicDevice represents a basic device instance.
type BasicDevice struct {
    // Attributes defines the set of attributes for this device.
    // The name of each attribute must be unique in that set.
    //
    // The maximum number of attributes and capacities combined is 32.
    //
    // +optional

>     Attributes map[QualifiedName]DeviceAttribute

Reviewer 7: Questions around the API design:
- Any specific reason for this name? For me, it'd be natural if we could call it `Labels` or `Annotations`, which is familiar names within K8s APIs. Is Attribute a more common name for wg/device-management domain?
- Any reason that we cannot do `map[string]string` (which you had before) and rather you prefer to do `map[QualifiedName]DeviceAttribute`? `QualifiedName` is fine. But, `DeviceAttribute` looks weird `boolValue: true` is unfamiliar API, and also maybe we'll have to get more fields to cover other types and stuff in the future, which could bring a troublesomeness here.

Reviewer: This is from core DRA. Device attributes are not all just strings, hence the `DeviceAttribute` one-of struct.

Reviewer 7: Okay, if both are coming from the existing concept, then that's fine

Reviewer: This field is no longer relevant for the KEP.

You can replace it with `...` to indicate that `BasicDevice` has some other fields besides the ones you add here.

The basic idea is the following:

1. **Adding Attributes to ResourceSlice**:
   - Add an attribute to `ResourceSlice` to indicate fabric devices. This key is predefined as part of the attributes.

2. **Waiting for Device Attachment in PreBind**:
   - For fabric devices, the scheduler waits for the device attachment to complete during the `PreBind` phase.

3. **PreBind Process**:
   The overall flow of the `PreBind` process is as follows:

   - **Updating ResourceClaim**:
     - The scheduler DRA plugin updates the `ResourceClaim` to notify the Composable DRA Controllers that device attachment is needed. 
     This is the same as the existing `PreBind` process.
     - In addition to the existing operations, the update to the `ResourceClaim` includes setting the necessary values in the `AllocatedDeviceStatus` conditions.

   - **Monitoring and Preparation by Composable DRA Controllers**:
     - Composable DRA Controllers monitor the `ResourceClaim`. If a device that requires preparation is associated with the `ResourceClaim`, they perform the necessary preparations.
     - Once the preparation is complete, they set the conditions to `true`.

   - **Completion of the PreBind Phase**:

>      - Once all conditions are met, the `PreBind` phase is completed, and the scheduler proceeds to the next step.

Reviewer 2: This sentence is not true. IIUC the plan is actually to abandon binding instead of 'proceed to the next step'.

Reviewer 6: In the case of a Composable system, scheduler abandon binding.
However, we are considering general-purpose use outside of the Composable system, so we should proceed to the Bind phase if all conditions are met in BindingGates.  
Therefore, in a Composable system, the Composable DRA Controller must intentionally change the conditions in BindingFailureGates to true after attach is complete. I will update KEP.

(Of course, we can leave it attached to that node for future use by the Pod.)

This issue needs to be resolved before the beta is released.

**Pods which are not bound yet (in api-server) and not unschedulable (in api-server) are not visible by cluster autoscaler, so there is a risk that the node will be turned down.**

Regarding collaboration with the Cluster Autoscaler, using node nomination can address the issue.
This issue needs to be resolved before the beta is released.

**The in-flight events cache may grow too large when waiting in PreBind.**

To address the PreBind concern, the solution is to modify the scheduling framework to flush the in-flight events cache before PreBind.
This prevents issues in the scheduling queue caused by keeping pods at PreBind for an extended period.
This issue will be addressed separately as outlined in kubernetes/kubernetes#129967.
This issue needs to be resolved before the beta is released.

## Design Details

### DRA Scheduler Plugin Design Overview

This document outlines the design of the DRA Scheduler Plugin, focusing on the handling of fabric devices.

> Key additions include `BindingConditions` and `BindingFailureConditions` for device identification and preparetion, enhancements to `AllocatedDeviceStatus`, and the process for handling `ResourceSlices` upon attachment failure.

Reviewer 5: s/preparetion/preparation/

Author: Thank you for your review. I have reflected all of your "requested changes" in the KEP.

For this feature, following fields are added:

```go
// AllocatedDeviceStatus contains the status of an allocated device, if the
// driver chooses to report it. This may include driver-specific information.
type AllocatedDeviceStatus struct {
    ...
    // BindingConditions defines the conditions for proceeding with binding. All listed conditions must be True to proceed with binding. 
    //
    // +optional
    BindingConditions []string

    // BindingFailureConditions defines the conditions for binding failure.
    // If any is True, a binding failure occurred.
    //
    // +optional
    BindingFailureConditions []string

    // UsageRestrictedToNode indicates if the usage of an allocation involving this device
    // has to be limited to exactly the node that was chosen when allocating the claim.
    //
    // +optional
    UsageRestrictedToNode bool

    // BindingTimeout indicates the prepare timeout period.
    // If the timeout period is exceeded before all BindingConditions reach a True state, the scheduler clears the allocation in the ResourceClaim and reschedules the Pod.
    //

>     // +optional

Reviewer: ```suggestion
    //  The default timeout if not set is 10 minutes.
    //
    // +optional
```

Author: Thank you for your review. I have reflected all of your suggestions and comments.

- [ ] (R) Graduation criteria is in place
  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
- [ ] (R) Production readiness review completed
- [ ] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

To achieve efficient management of fabric devices, we propose adding the following features to the Kubernetes scheduler's DRA plugin.

> Fabric devices are those that are not directly connected to the server and require attachment to the server for use.

Reviewer 8: wdym by server?
Did you mean node, cluster or something else?

Reviewer 2: IIUC it's a physical machine. Nodes are one or more VMs.

// driver chooses to report it. This may include driver-specific information.
type AllocatedDeviceStatus struct {
    ...
    // BindingGates defines the gates for binding.
    //
    // +optional
    BindingGates map[string]bool

    // BindingFailureGates defines the gates for binding failure.
    //
    // +optional
    BindingFailureGates map[string]bool
}
```

#### Scheduler DRA Plugin Modifications

When `UsageRestrictedToNode: true` is set, the scheduler DRA plugin will perform the following steps:

1. **Set NodeSelector**: Before the `PreBind` phase, add the `NodeName` to the `ResourceClaim`'s `NodeSelector`.

If Gates are present, the scheduler DRA plugin will perform the following steps during the `PreBind` phase:

2. **Copy Gates**: Copy `BindingGates` and `BindingFailureGates` from `ResourceSlice.Device.Basic` to `AllocatedDeviceStatus`.
3. **Wait for Conditions**: Wait for the following conditions:
   - If `NeedToPreparing` is `True`, wait until `IsPrepared` is `True` before proceeding to Bind.
   - If `PreparingFailed` is `True`, clear the allocation in the `ResourceClaim` and reschedule the Pod.

>    - If the preparation takes longer than the `PreparingTimeout` period, clear the allocation in the `ResourceClaim` and reschedule the Pod.

Reviewer 6: OK, I defined it in BasicDevice. And updated the KEP including your other comments about BindingGates

To support these steps, the following keys are defined:

```go

> const (

Reviewer 5: These are all now driver-defined and not K8s things. There should be no need for them here.

If you want to include an *example* of how driver's could use the new fields to manage an attachment process, that would be great. But it's an example only.

    NeedToPreparing = "kubernetes.io/need-to-preparing"

    // IsPrepared indicates the device ready state.
    // If NeedToPreparing is True and IsPrepared is True, the scheduler proceeds to Bind.
    IsPrepared = "kubernetes.io/is-prepared"

    // PreparingFailed indicates the device preparation failed state.
    // If PreparingFailed is True, the scheduler will clear the allocation in the ResourceClaim and reschedule the Pod.
    PreparingFailed = "kubernetes.io/preparing-failed"

    // PreparingTimeout indicates the prepare timeout period.
    // If the timeout period is exceeded, the scheduler clears the allocation in the ResourceClaim and reschedules the Pod.
    PreparingTimeout = "kubernetes.io/preparing-timeout"
)
```

Note: There is a concern that the in-flight events cache may grow too large when waiting in PreBind.
To address this, the scheduling framework will be modified to flush the in-flight events cache before PreBind.
This prevents issues in the scheduling queue caused by keeping pods at PreBind for an extended period.
This issue will be addressed separately as outlined in kubernetes/kubernetes#129967.

> #### PreBind Phase Timeout

Reviewer 5: We should decide if we need a hard timeout, or can just use whatever value is given to us. We can also consider having a default timeout if it's not present in the ResourceSlice.

A hard timeout puts a limit on the utility. I would probably prefer having a default, but allowing the vendor to specify an arbitrarily long timeout in the ResourceSlice. So long as that is not disruptive to other pods trying to schedule.

In other words, 5m could be a default, but not hard coded.

for the KEP to move to `implementable` status and be included in the release.

In some cases, the questions below should also have answers in `kep.yaml`. This
is to enable automation to verify the presence of the review, and to reduce review
burden and latency.

The KEP must have a approver from the
[`prod-readiness-approvers`](http://git.k8s.io/enhancements/OWNERS_ALIASES)
team. Please reach out on the
[#prod-readiness](https://kubernetes.slack.com/archives/CPNHUMN74) channel if
you need any help or guidance.
-->

### Feature Enablement and Rollback

<!--
This section must be completed when targeting alpha to a release.
-->

###### How can this feature be enabled / disabled in a live cluster?

<!--
Pick one of these and delete the rest.

Documentation is available on [feature gate lifecycle] and expectations, as
well as the [existing list] of feature gates.

[feature gate lifecycle]: https://git.k8s.io/community/contributors/devel/sig-architecture/feature-gates.md
[existing list]: https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/
-->

- [x] Feature gate (also fill in values in `kep.yaml`)

>   - Feature gate name: DRADeviceAttachDuringScheduling

Reviewer 5: I think this should be something like `DRAPrebindingGates`, or whatever similar name is needed based on the naming we choose above. The idea is to separate out the actual code change (add a success/failure/timeout wait during prebind), from the specific use case (attach devices).

-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

<!--
This section is incredibly important for producing high-quality, user-focused
documentation such as release notes or a development roadmap. It should be
possible to collect this information before implementation begins, in order to
avoid requiring implementors to split their attention between writing release
notes and implementing the feature itself. KEP editors and SIG Docs
should help to ensure that the tone and content of the `Summary` section is
useful for a wide audience.

A good summary is probably at least a paragraph in length.

Both in this section and below, follow the guidelines of the [documentation
style guide]. In particular, wrap lines to a reasonable length, to make it
easier for reviewers to cite specific portions, and to minimize diff churn on
updates.

[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md

> -->

Author: Thank you for your feedback. In the next update, I will address your comments, and remove them from all sections where have been filled out.

required) or even code snippets. If there's any ambiguity about HOW your
proposal will be implemented, this is the place to discuss them.
-->

### DRA scheduler plugin Design overview

Add a flag to the `Device` within `ResourceSlice` to indicate whether it represents a fabric device.
This flag will be used by the controller that exposes the `ResourceSlice` to notify whether the device is a fabric device.
To avoid impacting existing DRA functionality, the default value of this flag is set to `false`.

```go
// Device represents one individual hardware instance that can be selected based
// on its attributes. Besides the name, exactly one field must be set.
type Device struct {
	// Name is unique identifier among all devices managed by
	// the driver in the pool. It must be a DNS label.
	//
	// +required
	Name string

	// Basic defines one device instance.
	//
	// +optional
	// +oneOf=deviceType
	Basic *BasicDevice

	// FabricDevice represents whether this device is a fabric device or not.
	// If true, it indicates that the device is connected via a fabric network.
	// This flag helps in distinguishing fabric devices from other types of devices.
	//
	// +optional

> 	FabricDevice bool

Reviewer 5: do we need this? can this be an attribute?

Reviewer 5: Hmm. Suppose we can use `Conditions` as I suggest in my other comment. In that case, the allocator needs to know to set the `Ready` (or `Schedulable`? `Unschedulable`?) condition to false for these devices. Other devices either have no `Ready` condition or need it defaulted to `Yes`.

Here's a thought: what if we add a field `InitialConditions []Condition`? The allocator would copy that field to every allocated device upon creation, to populate the `Conditions` field. From that point on, the driver would own the `Conditions` field (as it does today).

I think that, along with the scheduler plugin honoring that new condition (or perhaps allowing something like `SchedulingGates` to require a list of conditions to be satisfied), would be enough to satisfy your use case.

Reviewer 5: Actually initial conditions may not be needed, just a list of device scheduling gates, which could be copied from the ResourceSlice to the `AllocationResult`. If present, the scheduler would not schedule until the driver populates those conditions and they reach `True` status. This leaves `AllocatedDeviceStatus` 100% driver owned.

Author: By reserving a specific key in the attribute and implementing the scheduler plugin to respect that attribute, I believe this field can be treated as an attribute.

As for the usage, as you mentioned in the comment, the allocator will check this attribute, initialize the AllocatedDevices upon device creation, and set the DeviceAttach condition. The driver will update the conditions set in the ResourceClaim to notify the scheduler plugin of the device attachment completion.

In this case, the role of the scheduler plugin is to look at the attribute in the ResourceSlice, initialize the AllocatedDevices, and wait for readiness. The role of the driver is to set the attribute when publishing the ResourceSlice, prepare it once it is bound to the ResourceClaim, and report the result.

Author: The most important point is that we reserve the key for this function in the attribute. If this is allowed, there is no problem in using the attribute.

Reviewer 5: Now that I understand better, `binding-gates` is probably a better name. The idea is that in the Device in the ResourceSlice, we add a field: `BindingGates []string`. That's the only change to ResourceSlice. This is used in `PreBind` to evaluate the `Conditions` in the `AllocatedDeviceStatus`. All listed condition names in `BindingGates` must be present and have the value `True` in order to pass PreBind. We don't need `InitialConditions` if we have this, so we avoid two agents writing to `AllocatedDeviceStatus`.

We need to know when to stop waiting for the conditions, too. That means one of these things, I think:
- A fixed time out value
- A timeout value included in the ResourceSlice's Device
- A well-known condition name to indicate unrecoverable failure of attachment
- A separate field in ResourceSlice with a condition name (or names) that indicate unrecoverable failure

Reviewer: That `BindingGates` contains the name of conditions isn't immediately obvious. `BindingConditions`?

This looks like a more generic variant of [what I had proposed](https://github.com/kubernetes/enhancements/pull/5012#discussion_r1935487560). It may indeed be useful.

> A well-known condition name to indicate unrecoverable failure of attachment

When we do that, the "generic" aspect gets lost a bit.

> A separate field in ResourceSlice with a condition name (or names) that indicate unrecoverable failure.

That is generic again. Can we use only this and avoid the well-known condition name?

Author: Let me clarify my understanding. If we introduce the binding-gate, it would work as follows:

1. The DRA composable controller writes the necessary conditions into the Device.binding-gate within the `ResourceSlice` and publishes the `ResourceSlice`.
2. The scheduler DRA plugin creates the `AllocatedDeviceStatus`. At this time, it writes the condition names from the binding-gate into the Conditions.
3. The DRA composable controller attaches the device and updates the Conditions in the `AllocatedDeviceStatus`.
4. The scheduler DRA plugin waits in `PreBind` until all Conditions are true.

As mentioned, it is also necessary to notify when to stop waiting for the Conditions.

Which is better: defining some conditions as DRA or allowing vendors to define specific values through the binding-gate? I feel that the three conditions [mentioned here](https://github.com/kubernetes/enhancements/pull/5012#discussion_r1935487560) can achieve the feature:

- Type: kubernetes.io/needs-attaching
- Type: kubernetes.io/is-attached
- Type: kubernetes.io/attach-failed

> We need to know when to stop waiting for the conditions, too. 

While I initially assumed a fixed timeout for stopping the wait for conditions, having a configurable timeout value that users or vendors can specify would be more versatile. I also think it would be beneficial to define a maximum timeout value as a fixed value.

Reviewer: I think @johnbelamaric was suggesting that the scheduler doesn't populate any conditions. It just waits for those to be `True` which are listed in a device's `BindingConditions` (my name, John used `BindingGates`).

The scheduler aborts once it sees a condition as `True` which is listed under `BindingFailures` (again my name, John only said "separate field").

Optionally we could pre-define a well-known binding failure condition, but I wasn't so sure about that because it breaks the "generic" aspect of this design.

Reviewer 5: Correct - there is no need for the scheduler to write any conditions. The schedule just looks at the binding gates field; if it is populated, then it waits for those to appear and be `True` before proceeding. If it is not populated, it works as it does today.

Similarly, a `BindingFailures` or `BindingFailureConditions` field probably should only be populated if there are binding gates. Should we *also* have a default timeout? Not sure if we should.

The name "binding gates" follows the same naming as [pod readiness gates](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-readiness-gate) and [pod scheduling gates](https://kubernetes.io/docs/concepts/scheduling-eviction/pod-scheduling-readiness/). I agree it's not clear those are conditions, but that's the way we do it elsewhere. Unfortunately those two features use the word "gates" but work in totally different ways...but this binding suggestion follows the pod readiness pattern.

I defer to the API reviewers for the right naming...

Reviewer: It occurred to me that "scheduler doesn't populate any conditions" leads to one problem: suppose a claim is allocated and the scheduler waits for the expected conditions. Now the scheduler restarts and the ResourceSlice with the device, for whatever reason, gets deleted. Now if the scheduler continues, it sees a claim that is ready for use and will use it without waiting. Also, when a user inspects the claim while things are pending it's not obvious what is going on.

The underlying reason is that it violates a design principle of DRA: an allocated claim needs to be usable and complete without any additional knowledge, like the class or slice that were used to allocate it.

To fix this, the AllocatedDeviceStatus would have to be extended with a copy of the BindingGates (aka BindingConditions) and BindingFailures from the time of the allocation.

At that point, being generic implies adding quite a few additional API fields, compared to reusing existing fields (attributes, conditions) with some pre-defined attribute and condition names. Is that worth it? I'm undecided.

@johnbelamaric: when you suggested `BindingGates`, did you have specific additional use cases in mind beyond "attaching"?

Author: @pohly @johnbelamaric 
Is my understanding correct that the current discussion revolves around the following two proposals?

1. **Proposal 1**: Prepare BindingGate and BindingFailureGates in the ResourceSlice's Device and copy them to AllocatedDeviceStatus (though this involves significant API extensions).
2. **Proposal 2**: Define several attribute names and status names related to device attachment and use them to manage the state (though this requires the scheduler to set `Conditions`).

From the perspective of functionality, I believe either proposal would work. Which design do you think the community is more likely to accept?

Reviewer: In https://github.com/kubernetes/enhancements/pull/5012#discussion_r1942196626 @johnbelamaric wrote (replying here because it pertains to "generic binding gates" vs. "dedicated API"):
> We should advertise the fabric device as available from multiple nodes, with the binding gates, failure conditions, and timeout. The scheduler picks a possible node and puts that in the node name of the allocation. 

There's a small gap here: if binding gates are opaque, the scheduler doesn't know that "binding gates exist" means "allocation must be local to a node". It would produce a claim status that replicates the "available on multiple nodes" node selector of the pool, which is wrong and doesn't ensure that other pods really land on the chosen node. It also hides from the driver where it needs to attach - see https://github.com/kubernetes/enhancements/pull/5012#discussion_r1935487560.

Reviewer: > Is my understanding correct that the current discussion revolves around the following two proposals?

Correct.

> Which design do you think the community is more likely to accept?

Uncertain :hot_face:

But note [the gap in the generic approach](https://github.com/kubernetes/enhancements/pull/5012#discussion_r1942431579) that I added in parallel to your reply: personally I am leaning towards "proposal 2" because it doesn't have that gap and is simpler (less new API fields) and easier to document (no need to talk about abstract binding gates without knowing what drivers are going to use).

Reviewer 6: > But note https://github.com/kubernetes/enhancements/pull/5012#discussion_r1942431579 that I added in parallel to your reply: personally I am leaning towards "proposal 2" because it doesn't have that gap and is simpler (less new API fields) and easier to document (no need to talk about abstract binding gates without knowing what drivers are going to use).

I prefer Proposal2 for the same reason as you.

@johnbelamaric 
Please let us know your opinion.

Reviewer: If you don't hear from John otherwise, then let's go with proposal 1, dedicated fields. Separate fields fit better into how Kubernetes defines and evolves APIs.

To close the gap above, we need another field in the `BasicDevice` which tells the scheduler to make the claim specific to the node - something like "UsageRestrictedToNode - if set, usage of an allocation involving this device has to be limited to exactly the node that was chosen when allocating the claim". As usual, name to be decided...

I'm intentionally not using the word "attach" because proposal 1 is supposed to be generic. This new field should describe the effect, not why it gets set.

Reviewer 6: OK. Let's think about Proposal 1.

I haven't thought of any use cases other than device attachment, but I believe that the key concept is "device that needs some preparation before binding a pod to a node."
So I thought of the following effects. (I think there is a more appropriate name, but for now, focus on the effect...)

- UsageRestrictedToNode
  Usage of an allocation involving this device has to be limited to exactly the node that was chosen when allocating the claim
- NeedToPreparing
  This device needs some preparation. If this flag is True, the scheduler waits in PreBind.
- IsPrepared
  The device ready state. If NeedToPreparing is True and IsPrepared is True, the scheduler proceeds to Bind.
- PreparingFailed
  The device preparation failed state. If PreparingFailed is True, the scheduler will clear the allocation in the ResourceClaim and reschedule the Pod.
- PreparingTimeout
  The prepare timeout period. If the timeout period is exceeded, the scheduler clears the allocation in the ResourceClaim and reschedules the Pod.

If this is reasonable, I will update the KEP.

}
```

**Additions to the DRA Scheduler Plugin**

> In the current implementation, the `PreBind` phase waits until the `ResourceClaim` update is completed.

Reviewer: I'd like to hear from SIG Scheduling folks whether this key design idea is viable.

The implication is that the kube-scheduler will have pods stuck in `PreBind` for longer and long-running goroutines. Go should handle that fine, but I am a bit concerned about the effect on queuing and caches maintained by the scheduler.

cc @sanposhiho @alculquicondor

Reviewer 2: My first thought is that waiting in the pre-bind is feasible and preferred way of handling such dynamic allocation. It's because scheduler internally considers scheduling as done and does not expect changes unless binding fails (a legitimate scenario here) or pod-in-binding gets preempted. So I'd focus on thinking how the preemption of in-binding and after-binding cases should look like, as it's completely unclear to me now.

One risk with delayed pre-binding is that there is a known issue that pods which are not bound yet (in api-server) and not unschedulable (in api-server) are not visible by cluster autoscaler, so there is a risk that the node will be turned down. We however plan a remedy here (we will use node nomination mechanism to communicate this decision), so no need to worry about this issue.

So from scheduling perspective, the alternative approach would be to reschedule the pod (proposal2) and make it unschedulable until the device is bound. The problem here is that there is no good mechanism which would guarantee that other node won't jump into the node in the meantime. There is indeed the node nomination mechanism which scheduler could just use internally (without talking to api-server during scheduling cycle), but so far it was used only to reserve a place for preempting pod until the preempted pods gets removed. I was also thinking to use this mechanism more widely in scheduler, but to use it, we'd most likely need to perform bigger changes in scheduler. The main problem for now is that nominations are entirely used for preemption and setting it by plugins may break other invariants.

For instance if other priority pod needs to use a node which is already nominated, it can just use it, without initiating the preemption process, so in our case w could prevent from lower or equal priority pods to use the space, but higher priority pods could still jump in even though the allocation process has started in the meantime.

So I'm currently interested the most how the preemption (or node hijacking case) could work with this dynamic allocation. In particular, is scheduler able to simulate freeing the fabric devices in its default preemption analysis which trigger each time there are no available resources for new pods.

Reviewer 10: One thing that comes to my mind is inFlightEvents list in scheduler. This would grow up significantly, storing all (important) cluster events that happened between popping the pod from queue (start of scheduling) and finishing binding. Aren't we able to put a pod into unschedulable queue while we wait for the device to be attached?

Reviewer 2: The list of events is indeed cleared after pre-bind just before bind, but on the other hand don't we have this problem also if we consider the pod unschedulable?

Author: @macsko  @dom4ha 
Thank you for the review from the scheduler's perspective.

> My first thought is that waiting in the pre-bind is a feasible and preferred way of handling such dynamic allocation.

I'm glad to hear that the idea of waiting for device attachment in the PreBind phase is fundamentally acceptable.

On the other hand, I understand there are concerns related to preemption. I had assumed that Pods in the PreBind phase of the binding cycle were not subject to preemption. Is that not the case?

> So from scheduling perspective, the alternative approach would be to reschedule the pod (proposal2) and make it unschedulable until the device is bound. The problem here is that there is no good mechanism which would guarantee that other node won't jump into the node in the meantime. There is indeed the node nomination mechanism which scheduler could just use internally (without talking to api-server during scheduling cycle), but so far it was used only to reserve a place for preempting pod until the preempted pods gets removed. I was also thinking to use this mechanism more widely in scheduler, but to use it, we'd most likely need to perform bigger changes in scheduler. The main problem for now is that nominations are entirely used for preemption and setting it by plugins may break other invariants.

I am not fully understanding this paragraph, so could you please provide a bit more explanation? When rescheduling a Pod, is it necessary to nominate a node?

>So I'm currently interested the most how the preemption (or node hijacking case) could work with this dynamic allocation. In particular, is scheduler able to simulate freeing the fabric devices in its default preemption analysis which trigger each time there are no available resources for new pods

Even if this node hijack case occurs due to preemption, I believe the status of the fabric devices will be responsibly published as ResourceSlices by the vendor's controller. By using this information for scheduling, I think consistency can be maintained even with preemption. Or is this concern related to other problematic cases?

Additionally, are there any other concerns besides preemption? If the preemption concerns are resolved, would the review from the scheduler's perspective be considered LGTM?

Reviewer 10: > On the other hand, the concern with the proposal2 is that the following scheduling cycle may not let the pod to get scheduled in the previously selected node because some other pod could get scheduled in the meantime.

That's right. I think, for v1.33 we won't be able to do the approach with requeueing (it would require changing the semantics of nominatedNodeName). So, I think PreBind approach is better at the moment, but we need to make sure the memory consumption won't be a problem. 

@KobayashiD27 do we know what would be the typical waiting time for the device to be attached?

Reviewer 2: Ideally any sort of physical allocation (I know this word is a bit overloaded) should always happen in the binding phase, but the scheduling phase should work only on logical allocations/reservations.

Such distinction would be useful in tracking of physical resources deallocation whenever the binding process fails or a pod is preempted before the binding is completed. Obviously, we should avoid reverting such decisions as much as possible, as they are disruptive to the system, but this is exactly what this KEP is trying to achieve.

In contrast, the scheduling phase could change the scheduling decision without worrying that any cleanup actions need to be taken. In that case scheduler would perform any corrective actions in the following scheduling cycles, so "at some point" (eventual consistency) vs "right after the binding fails". This is how node nomination currently works.

I'd rather think how to make the binding phase refactored to avoid active polling in go routines in the future. In short term, I'd only eliminate the need for caching cluster events, but not sure how to achieve it yet (moving the discussion into a separate issue https://github.com/kubernetes/kubernetes/issues/129967).

Reviewer 7: I just left a comment on Dominik's issue. So, regarding the concern of PreBind, I think we can just change the scheduling framework side to flush the in-flight events cache before PreBind so that keeping pods at PreBind for a long time doesn't cause 💥 at the scheduling queue.

Reviewer 7: @KobayashiD27 
Please mention it somewhere, like we have this concern with the current scheduler, but we will solve in that way.

Author: @dom4ha 
> In any of the two alternatives, scheduler can always "preempt" lower priority pods and it can do it in a few different ways, for instance ignore the mentioned NominatedNodeName and schedule other higher priority pod instead. I'm not sure how much the preemption is a concern here, but I'm missing a part in this KEP saying what's going to happen if a pod fails to bind for whatever reason. Is there a timeout after which the device will get detached from the selected node?

If the device attachment fails, the PreBind will fail. We will also set a timeout for waiting in PreBind to prevent it from waiting indefinitely. In case of attachment failure or timeout, we plan to unbind the device from the ResourceClaim. It is also desirable to unbind the device from the ResourceClaim if the binding fails for reasons other than DRA. (Is there an appropriate place to handle this?) After the device is unbound from the ResourceClaim, the device will typically be detached from the selected node, depending on the vendor's driver implementation.

@macsko 
> Do we know what would be the typical waiting time for the device to be attached?

If the device attachment is successful, we expect it to take no longer than 5 minutes. Therefore, if we set a fixed timeout for the scheduler, we would like to set it to around 10 minutes. Is this feasible?

@sanposhiho 
> Please mention it somewhere, like we have this concern with the current scheduler, but we will solve in that way.

Based on this discussion, we will wait for the attachment in PreBind. There is a concern that the in-flight events cache may grow too large when waiting in PreBind, but we plan to solve this issue as outlined in  kubernetes/kubernetes#129967.
I will update the KEP to include this concern and our solution. Please let me know if there are any inaccuracies.

Reviewer 7: > Is there an appropriate place to handle this?

You can use [`Unreserve`](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/framework/interface.go#L630-L635) for unbinding. Unreserve is called at any failure within the binding cycle

This proposal adds functionality to block the completion of the `PreBind` phase until the device is attached if a fabric device is included in the `ResourceClaim`.

To communicate the completion of fabric device attachment to the scheduler, a flag will be added to the `Status` of the `ResourceClaim`.

```go
// AllocatedDeviceStatus contains the status of an allocated device, if the
// driver chooses to report it. This may include driver-specific information.
type AllocatedDeviceStatus struct {
...
	// DeviceAttached represents whether the device has been successfully attached.
	//
	// +optional

> 	DeviceAttached string

Author: Thank you for your suggestions. If there is a better implementation location, I would appreciate your guidance. I also believe that utilizing the existing Conditions, if feasible, is the best approach. I will check if Conditions can handle this and make the KEP changes if there are no issues.

Regarding the Nodename, is it possible to use the Data raw extension to communicate the Node decided by the scheduler to the driver? 

If we can achieve this functionality without API changes, that would be ideal.

> 	DeviceAttached string

Author: My understanding is that this feature waits for a device to be attached in the PreBind() phase.
In the PreBind() phase, I understand that one node is already selected.

> 	DeviceAttached string

Author: >The explanation for Nodename is a bit misleading: instead of "node where the device is attached" it should say "node where the device needs to be attached", shouldn't it?

The explanation of NodeName was indeed misleading. "Node where the device needs to be attached" more accurately describes the role of NodeName.

> So `kubernetes.io/needs-attaching: true` has to enable several things:
> 
> * When composing the node selector of an `AllocationResult`, match exactly one node.
> * Add a `AllocatedDeviceStatus` with a condition with `Type: kubernetes.io/needs-attaching` and `Status: True`. `Conditions` supports different owners for different entries with `Type` as key (relevant for server-side-apply), so I prefer to let the scheduler own a different entry than the driver. Having this condition provides some additional observability.
> * Wait for a condition with `Type: kubernetes.io/is-attached` and `Status: True` in `PreBind` before proceeding.
> * Give up when seeing a condition with `Type: kubernetes.io/is-attached` and Status: False`. It's debatable whether this should be a separate `kubernetes.io/attach-failed`.
> 
> With this definition, the driver can determine where to attach by parsing the node selector. I don't like that, because there are multiple different node selectors (by field, by label) which all would be valid. It might be better to add a separate field that stores the node more explicitly.
 
If we proceed with the approach of using existing `Conditions`, the idea of defining several ConditionTypes for this feature seems promising. By effectively handling these `Conditions`, we can achieve the functionality of completing the Pod's schedule after attaching the fabric device to the node. 

I believe having two types of ConditionTypes would suffice. 
(What are the advantages of separating `Type: kubernetes.io/is-attached`?)

> 	DeviceAttached string

Reviewer: Did you mean `kubernetes.io/is-attached` or `kubernetes.io/attach-failed`?

The main advantage of `kubernetes.io/is-attached` literally is that it's owned by the driver, and only by the driver.

The advantage of `kubernetes.io/attach-failed` is that `kubernetes.io/is-attached: False` doesn't clearly articulate the "permanently failed" aspect. It could also just be not attached yet, so making the scheduler give up and try elsewhere when it sees `kubernetes.io/is-attached: False` is a stretch.

> 	DeviceAttached string

Author: Sorry, I made a mistake. I was asking about the benefits of kubernetes.io/attach-failed. Also, I understood the reason for splitting the two types. Indeed, I think a type to represent persistent failures is necessary.

>     // BindingGates defines the gates for binding.

Reviewer 6: I changed BindingGates to BindingConditinos. I also reflected your other comments in the KEP.

	// NodeName contains the name of the node where the device is attached.
	//
	// +optional
	Nodename string 
}

const (
	DeviceAttachReady		= "Ready"
	DeviceAttachPreparing	= "Preparing"
	DeviceAttachFailed		= "Failed"
	DeviceAttachReschedule	= "Reschedule"

> )

Author: Here is a rough idea of the current state-flow we are considering: 
![image](https://github.com/user-attachments/assets/159ff42f-30a4-4bb1-817b-8d96fbfcd528)

After the bind is cleared, PreBind fails, and rescheduling occurs.

>Can the Preparing and Reschedule states be merged? I understand that the "Preparing" and "Reschedule" processes should do the same thing? Or do they need to be separated?
Is "Failed" the final state or is it still a process state? If the ResourceClaim attach process fails, will it change to Failed and then back to Reschedule?

Preparing and Reschedule cannot be merged.
Failed and Reschedule have the same behavior except for log output, so they can be merged in terms of state-flow.
However, we would like to distinguish between Failed and Reschedule for logging purposes. We would appreciate the community's opinion on this.
Please note that [this discussion](https://github.com/kubernetes/enhancements/pull/5012#discussion_r1935487560), so the handling of states may change in the future.

>     BindingTimeout int

Reviewer: To avoid ambiguity, other durations are explicitly called `Seconds`, so `BindingTimeoutSeconds`.

Or use `BindingTimeout *metav1.Duration`. Probably better.

For this feature, following fields are added:

```go
// AllocatedDeviceStatus contains the status of an allocated device, if the
// driver chooses to report it. This may include driver-specific information.
type AllocatedDeviceStatus struct {
    ...
    // BindingGates defines the gates for binding.
    //
    // +optional

>     BindingGates map[string]bool

Reviewer: I thought the idea was to use the existing `AllocatedDeviceStatus.Conditions` field.

Reviewer 5: This is here to provide the "as-was" list of gates (the list of gates as they were defined in the slice at the time of the allocation) - I think it should just be a `[]string`

Reviewer: Okay, now I get it. More comments on the field definitely would have helped.

Ack on `[]string`.

    // BindingFailureGates defines the gates for binding failure.
    //
    // +optional

>     BindingFailureGates map[string]bool

Reviewer: Same here.

Reviewer 5: same, `[]string`

}
```

#### Scheduler DRA Plugin Modifications

When `UsageRestrictedToNode: true` is set, the scheduler DRA plugin will perform the following steps:

1. **Set NodeSelector**: Before the `PreBind` phase, add the `NodeName` to the `ResourceClaim`'s `NodeSelector`.

If Gates are present, the scheduler DRA plugin will perform the following steps during the `PreBind` phase:

2. **Copy Gates**: Copy `BindingGates` and `BindingFailureGates` from `ResourceSlice.Device.Basic` to `AllocatedDeviceStatus`.
3. **Wait for Conditions**: Wait for the following conditions:
   - Wait until all conditions in the BindingGates are `True` before proceeding to Bind.
   - If any one of the conditions in the BindingFailureGates becomes `True`, clear the allocation in the `ResourceClaim` and reschedule the Pod.
   - If the preparation of a device takes longer than the `BindingTimeout` period, clear the allocation in the `ResourceClaim` and reschedule the Pod.

To support these steps, for example, a DRA driver can include the following definitions in BindingGates or BindingFailureGates within a ResourceSlice:

```go
const (
    // NeedToPreparing indicates that this device needs some preparation.

>     NeedToPreparing = "kubernetes.io/need-to-preparing"

Reviewer: I agree with https://github.com/kubernetes/enhancements/pull/5012/files#r1949713686, this is just an example. Because it's specific to a driver, it shouldn't use the `kubernetes.io` domain. You can use `dra.example.com` instead.

Reviewer: About "need-to-preparing": I don't see that being used. The binding gate (or rather, condition) is `is-prepared`.

)
```

#### AllocatedDeviceStatus Enhancements

The `BindingConditions` and `BindingFailureConditions` fields within `AllocatedDeviceStatus` are used to indicate the status of the device attachment.
These fields will contain a list of conditions, each representing a specific state or event related to the device.

For this feature, following fields are added:

```go
// AllocatedDeviceStatus contains the status of an allocated device, if the
// driver chooses to report it. This may include driver-specific information.
type AllocatedDeviceStatus struct {
    ...

>     // BindingConditions defines the conditions for binding.

Reviewer 5: ```suggestion
    // BindingConditions defines the conditions for proceeding with binding. All listed conditions must be True to proceed with binding. 
```

    //
    // +optional
    BindingConditions []string

    // BindingFailureConditions defines the conditions for binding failure.

>     // If true, a binding failure occurred.

Reviewer 5: s/If true/If any is True/

    //
    // +optional

>     BindingFailureConditions []stirng

Reviewer 5: s/stirng/string/

    // UsageRestrictedToNode indicates if the usage of an allocation involving this device
    // has to be limited to exactly the node that was chosen when allocating the claim.
    //
    // +optional
    UsageRestrictedToNode bool

    // BindingTimeoutSeconds indicates the prepare timeout period.

>     // If the timeout period is exceeded, the scheduler clears the allocation in the ResourceClaim and reschedules the Pod.

Reviewer 5: ```suggestion
    // If the timeout period is exceeded before all BindingConditions reach a True state, the scheduler clears the allocation in the ResourceClaim and reschedules the Pod.
```

    //
    // +optional

>     BindingTimeoutSeconds *metav1.Duration

Reviewer 5: It shouldn't be called `Seconds` if it's a Duration - it should just be `BindingTimeout` in that case.
